// @metamask/snaps-simulator/dist/esm/features/builder/components/BaseNode.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex, Text } from "@chakra-ui/react";
import { assert } from "@metamask/utils";
import { Icon } from "../../../components";
export var BaseNode = function(param) {
    var node = param.node, isDragging = param.isDragging, children = param.children, onClose = param.onClose;
    var _node_data;
    assert((_node_data = node.data) === null || _node_data === void 0 ? void 0 : _node_data.type, "Node must have a type.");
    var handleClick = function() {
        var _onClose;
        (_onClose = onClose) === null || _onClose === void 0 ? void 0 : _onClose(node);
    };
    return /*#__PURE__*/ _jsxs(Flex, {
        paddingY: "2",
        paddingLeft: "4",
        paddingRight: "2",
        background: "background.alternative",
        borderRadius: "lg",
        gap: "4",
        alignItems: "center",
        border: "1px solid",
        borderColor: "border.default",
        display: isDragging ? "none" : "flex",
        marginX: "4",
        cursor: node.id > 1 ? "move" : "default",
        children: [
            /*#__PURE__*/ _jsx(Icon, {
                icon: node.data.type,
                width: "16px"
            }),
            /*#__PURE__*/ _jsx(Text, {
                fontWeight: "500",
                fontSize: "sm",
                textTransform: "capitalize",
                color: "text.muted",
                display: "inline-block",
                minWidth: "65px",
                children: node.data.type
            }),
            children,
            node.id >= 2 && /*#__PURE__*/ _jsx(Icon, {
                icon: "cross",
                width: "11px",
                marginX: "0.5",
                marginLeft: "auto",
                cursor: "pointer",
                onClick: handleClick
            })
        ]
    });
};
