// @metamask/snaps-simulator/dist/esm/features/builder/components/Start.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex, Heading, Text } from "@chakra-ui/react";
import { Icon } from "../../../components";
export var Start = function() {
    return /*#__PURE__*/ _jsxs(Flex, {
        flexDirection: "column",
        alignItems: "center",
        borderColor: "border.default",
        padding: "4",
        paddingTop: "12",
        borderRadius: "lg",
        background: "background.alternative",
        flex: "1",
        marginY: "-4",
        children: [
            /*#__PURE__*/ _jsx(Icon, {
                icon: "drag",
                marginBottom: "1.5"
            }),
            /*#__PURE__*/ _jsx(Heading, {
                as: "h3",
                fontSize: "sm",
                fontWeight: "700",
                color: "text.muted",
                marginBottom: "1",
                children: "Drag components in here"
            }),
            /*#__PURE__*/ _jsxs(Text, {
                fontSize: "xs",
                fontWeight: "500",
                color: "text.muted",
                textAlign: "center",
                children: [
                    "Build your UI piece-by-piece using the",
                    /*#__PURE__*/ _jsx("br", {}),
                    "predefined components above."
                ]
            })
        ]
    });
};
