// @metamask/snaps-simulator/dist/esm/features/builder/components/TemplateComponentList.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Flex, List, ListItem, Text } from "@chakra-ui/react";
import { copyable, divider, heading, panel, text } from "@metamask/snaps-ui";
import { TemplateComponent } from "./TemplateComponent";
var TEMPLATE_COMPONENTS = [
    {
        icon: "panel",
        text: "Panel",
        data: panel([]),
        droppable: true
    },
    {
        icon: "heading",
        text: "Heading",
        data: heading("Heading"),
        droppable: false
    },
    {
        icon: "text",
        text: "Text",
        data: text("Text"),
        droppable: false
    },
    {
        icon: "divider",
        text: "Divider",
        data: divider(),
        droppable: false
    },
    {
        icon: "copyable",
        text: "Copyable",
        data: copyable("Copyable text"),
        droppable: false
    }
];
export var TemplateComponentList = function(param) {
    var nextId = param.nextId, incrementId = param.incrementId;
    return /*#__PURE__*/ _jsxs(Box, {
        marginBottom: "4",
        children: [
            /*#__PURE__*/ _jsx(Text, {
                fontSize: "xs",
                fontWeight: "600",
                lineHeight: "133%",
                marginBottom: "1",
                children: "Components"
            }),
            /*#__PURE__*/ _jsx(Flex, {
                as: List,
                gap: "2",
                children: TEMPLATE_COMPONENTS.map(function(component) {
                    return /*#__PURE__*/ _jsx(ListItem, {
                        children: /*#__PURE__*/ _jsx(TemplateComponent, {
                            incrementId: incrementId,
                            icon: component.icon,
                            node: {
                                id: nextId,
                                parent: 0,
                                droppable: component.droppable,
                                text: component.text,
                                data: component.data
                            }
                        })
                    }, "component-".concat(component.text));
                })
            })
        ]
    });
};
