// @metamask/snaps-simulator/dist/cjs/features/handlers/transactions/components/TransactionPrefills.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "TransactionPrefills", {
    enumerable: true,
    get: function() {
        return TransactionPrefills;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("react");
var _react1 = require("@chakra-ui/react");
var _presets = require("../presets");
var _TransactionPrefill = require("./TransactionPrefill");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
var TransactionPrefills = function(param) {
    var onClick = param.onClick;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react1.Box, {
        marginBottom: "4",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react1.Text, {
                fontWeight: "500",
                fontSize: "xs",
                marginBottom: "1",
                children: "Transaction presets"
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react1.Flex, {
                gap: "2",
                children: _presets.TRANSACTION_PRESETS.map(function(param, index) {
                    var name = param.name, transaction = param.transaction;
                    return /*#__PURE__*/ (0, _react.createElement)(_TransactionPrefill.TransactionPrefill, _object_spread_props(_object_spread({
                        name: name
                    }, transaction), {
                        key: "transaction-prefill-".concat(index),
                        onClick: onClick
                    }));
                })
            })
        ]
    });
};
