// @airstack/node/dist/esm/utils/addPageInfoFields.js
import { createPginationFieldAlias } from './cursor';
function getPageInfo(cursorName) {
    return {
        kind: 'Field',
        alias: {
            kind: 'Name',
            value: createPginationFieldAlias(cursorName),
        },
        name: {
            kind: 'Name',
            value: 'pageInfo',
        },
        selectionSet: {
            kind: 'SelectionSet',
            selections: [
                {
                    kind: 'Field',
                    name: {
                        kind: 'Name',
                        value: 'prevCursor',
                    },
                },
                {
                    kind: 'Field',
                    name: {
                        kind: 'Name',
                        value: 'nextCursor',
                    },
                },
            ],
        },
    };
}
export function addPageInfoFields(query, cursorName) {
    var _a;
    const fields = (_a = query === null || query === void 0 ? void 0 : query.selectionSet) === null || _a === void 0 ? void 0 : _a.selections;
    if (fields) {
        // eslint-disable-next-line
        // @ts-ignore
        fields.push(getPageInfo(cursorName));
    }
}
