// @metamask/snaps-simulator/dist/esm/features/layout/Layout.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex } from "@chakra-ui/react";
import { Outlet } from "react-router-dom";
import { Console } from "../console";
import { Header, Sidebar } from "./components";
/**
 * Render the layout of the application.
 *
 * @returns A React component.
 */ export var Layout = function() {
    return /*#__PURE__*/ _jsxs(Flex, {
        direction: "column",
        height: "100vh",
        children: [
            /*#__PURE__*/ _jsx(Header, {}),
            /*#__PURE__*/ _jsxs(Flex, {
                direction: "row",
                flex: "1",
                overflow: "hidden",
                children: [
                    /*#__PURE__*/ _jsx(Sidebar, {}),
                    /*#__PURE__*/ _jsxs(Flex, {
                        direction: "column",
                        width: "full",
                        children: [
                            /*#__PURE__*/ _jsx(Flex, {
                                flex: "1",
                                overflow: "hidden",
                                children: /*#__PURE__*/ _jsx(Outlet, {})
                            }),
                            /*#__PURE__*/ _jsx(Console, {})
                        ]
                    })
                ]
            })
        ]
    });
};
