// @metamask/snaps-simulator/dist/cjs/features/builder/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    getNodeText: function() {
        return getNodeText;
    },
    nodeModelsToComponent: function() {
        return nodeModelsToComponent;
    },
    panelToCode: function() {
        return panelToCode;
    }
});
var _snapsui = require("@metamask/snaps-ui");
var _snapsutils = require("@metamask/snaps-utils");
var _utils = require("@metamask/utils");
var _parsertypescript = /*#__PURE__*/ _interop_require_default(require("prettier/parser-typescript"));
var _standalone = /*#__PURE__*/ _interop_require_default(require("prettier/standalone"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function getNodeText(nodeModel) {
    (0, _utils.assert)(nodeModel.data, "Node model must have data.");
    if ((0, _utils.hasProperty)(nodeModel.data, "value")) {
        return nodeModel.data.value;
    }
    return null;
}
function nodeModelsToComponent(nodeModels) {
    var _root_data, _root;
    // We want to clone the node models so that we don't mutate the original data.
    var clonedModels = (0, _snapsutils.deepClone)(nodeModels);
    var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
    try {
        var _loop = function() {
            var nodeModel = _step.value;
            (0, _utils.assert)(nodeModel.data, "Node model must have data.");
            var parent = clonedModels.find(function(model) {
                return model.id === nodeModel.parent;
            });
            if (parent) {
                var _parent_data;
                (0, _utils.assert)(((_parent_data = parent.data) === null || _parent_data === void 0 ? void 0 : _parent_data.type) === "panel", "Parent must be a panel.");
                parent.data.children.push(nodeModel.data);
            }
        };
        for(var _iterator = clonedModels[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true)_loop();
    } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
    } finally{
        try {
            if (!_iteratorNormalCompletion && _iterator.return != null) {
                _iterator.return();
            }
        } finally{
            if (_didIteratorError) {
                throw _iteratorError;
            }
        }
    }
    var root = clonedModels.find(function(model) {
        return model.parent === 0;
    });
    (0, _utils.assert)(((_root = root) === null || _root === void 0 ? void 0 : (_root_data = _root.data) === null || _root_data === void 0 ? void 0 : _root_data.type) === "panel", "Root must be a panel.");
    return root.data;
}
/**
 * Get all the component types used in a component.
 *
 * @param component - The component.
 * @returns The component types.
 */ function getComponentTypes(component) {
    var componentTypes = new Set();
    componentTypes.add(component.type);
    var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
    try {
        for(var _iterator = component.children[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true){
            var child = _step.value;
            componentTypes.add(child.type);
            if (child.type === "panel") {
                var childComponentTypes = getComponentTypes(child);
                var _iteratorNormalCompletion1 = true, _didIteratorError1 = false, _iteratorError1 = undefined;
                try {
                    for(var _iterator1 = childComponentTypes[Symbol.iterator](), _step1; !(_iteratorNormalCompletion1 = (_step1 = _iterator1.next()).done); _iteratorNormalCompletion1 = true){
                        var childComponentType = _step1.value;
                        componentTypes.add(childComponentType);
                    }
                } catch (err) {
                    _didIteratorError1 = true;
                    _iteratorError1 = err;
                } finally{
                    try {
                        if (!_iteratorNormalCompletion1 && _iterator1.return != null) {
                            _iterator1.return();
                        }
                    } finally{
                        if (_didIteratorError1) {
                            throw _iteratorError1;
                        }
                    }
                }
            }
        }
    } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
    } finally{
        try {
            if (!_iteratorNormalCompletion && _iterator.return != null) {
                _iterator.return();
            }
        } finally{
            if (_didIteratorError) {
                throw _iteratorError;
            }
        }
    }
    return Array.from(componentTypes).sort(function(a, b) {
        return a.localeCompare(b);
    });
}
/**
 * Get the arguments used to create a component.
 *
 * @param component - The component.
 * @returns The arguments.
 */ function getComponentArgs(component) {
    switch(component.type){
        case _snapsui.NodeType.Panel:
            return component.children.map(getComponentArgs).join(",\n");
        case _snapsui.NodeType.Text:
        case _snapsui.NodeType.Heading:
        case _snapsui.NodeType.Copyable:
            return "'".concat(component.value, "'");
        case _snapsui.NodeType.Spinner:
        case _snapsui.NodeType.Divider:
        default:
            return "";
    }
}
/**
 * Get the code for a component.
 *
 * @param component - The component.
 * @returns The code.
 */ function componentToCode(component) {
    if (component.type === _snapsui.NodeType.Panel) {
        return "panel([\n".concat(component.children.map(componentToCode).join(",\n"), "\n])");
    }
    var args = getComponentArgs(component);
    return "".concat(component.type, "(").concat(args, ")");
}
function panelToCode(component) {
    var types = getComponentTypes(component).join(", ");
    return _standalone.default.format("\n      import { ".concat(types, " } from '@metamask/snaps-ui';\n\n      const component = ").concat(componentToCode(component), ";\n"), {
        parser: "typescript",
        plugins: [
            _parsertypescript.default
        ],
        printWidth: 80,
        tabWidth: 2,
        singleQuote: true,
        trailingComma: "all"
    });
}
