// @metamask/snaps-simulator/dist/esm/components/Author.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex, Text } from "@chakra-ui/react";
import { SnapIcon } from "./SnapIcon";
export var Author = function(param) {
    var snapName = param.snapName, snapId = param.snapId;
    return /*#__PURE__*/ _jsxs(Flex, {
        gap: "2",
        marginX: "4",
        marginY: "4",
        border: "1px solid",
        borderColor: "border.default",
        borderRadius: "32px",
        padding: "1",
        alignItems: "center",
        children: [
            /*#__PURE__*/ _jsx(SnapIcon, {
                snapName: snapName
            }),
            /*#__PURE__*/ _jsxs(Flex, {
                direction: "column",
                justify: "center",
                children: [
                    /*#__PURE__*/ _jsx(Text, {
                        fontSize: "sm",
                        lineHeight: "157%",
                        fontFamily: "custom",
                        children: snapName
                    }),
                    /*#__PURE__*/ _jsx(Text, {
                        fontSize: "xs",
                        lineHeight: "167%",
                        fontFamily: "custom",
                        color: "text.alternative",
                        children: snapId
                    })
                ]
            })
        ]
    });
};
