// @metamask/snaps-simulator/dist/esm/components/Delineator.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Flex, Text } from "@chakra-ui/react";
import { Icon } from "./Icon";
export var Delineator = function(param) {
    var snapName = param.snapName, children = param.children;
    return /*#__PURE__*/ _jsxs(Box, {
        border: "1px solid",
        borderColor: "border.default",
        borderRadius: "md",
        children: [
            /*#__PURE__*/ _jsxs(Flex, {
                direction: "row",
                alignItems: "center",
                padding: "1.5",
                borderBottom: "1px solid",
                borderColor: "border.default",
                children: [
                    /*#__PURE__*/ _jsx(Icon, {
                        icon: "snap",
                        width: "16px",
                        marginRight: "1"
                    }),
                    /*#__PURE__*/ _jsxs(Text, {
                        fontFamily: "custom",
                        fontSize: "xs",
                        children: [
                            "Content from ",
                            snapName
                        ]
                    })
                ]
            }),
            /*#__PURE__*/ _jsx(Box, {
                padding: "3",
                children: children
            })
        ]
    });
};
