// @metamask/snaps-simulator/dist/esm/components/Root.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ChakraProvider, ColorModeScript } from "@chakra-ui/react";
import { getBackendOptions, MultiBackend } from "@minoru/react-dnd-treeview";
import { StrictMode } from "react";
import { DndProvider } from "react-dnd";
import { Provider } from "react-redux";
import { Notifications } from "../features";
import { theme } from "../theme";
/**
 * Render the root component. This is the top-level component that wraps the
 * entire application.
 *
 * @param props - The props.
 * @param props.store - The Redux store.
 * @param props.children - The children to render.
 * @returns The root component.
 */ export var Root = function(param) {
    var store = param.store, children = param.children;
    return /*#__PURE__*/ _jsx(StrictMode, {
        children: /*#__PURE__*/ _jsx(Provider, {
            store: store,
            children: /*#__PURE__*/ _jsx(ChakraProvider, {
                theme: theme,
                children: /*#__PURE__*/ _jsxs(DndProvider, {
                    backend: MultiBackend,
                    options: getBackendOptions(),
                    children: [
                        /*#__PURE__*/ _jsx(Notifications, {}),
                        /*#__PURE__*/ _jsx(ColorModeScript, {
                            initialColorMode: theme.config.initialColorMode
                        }),
                        children
                    ]
                })
            })
        })
    });
};
