// @metamask/snaps-simulator/dist/esm/components/Window.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex } from "@chakra-ui/react";
import { Author } from "./Author";
/**
 * A MetaMask-like window, with a snap authorship pill.
 *
 * @param props - The props.
 * @param props.snapName - The name of the snap.
 * @param props.snapId - The ID of the snap.
 * @param props.children - The children to render inside the window.
 * @param props.showAuthorship - Show the authorship component.
 * @returns The window component.
 */ export var Window = function(param) {
    var snapName = param.snapName, snapId = param.snapId, children = param.children, _param_showAuthorship = param.showAuthorship, showAuthorship = _param_showAuthorship === void 0 ? true : _param_showAuthorship;
    return /*#__PURE__*/ _jsxs(Flex, {
        direction: "column",
        boxShadow: "lg",
        maxWidth: "360px",
        height: "600px",
        paddingY: "4",
        overflowY: "scroll",
        children: [
            showAuthorship && /*#__PURE__*/ _jsx(Author, {
                snapName: snapName,
                snapId: snapId
            }),
            children
        ]
    });
};
