// @metamask/snaps-simulator/dist/cjs/components/dialogs/ConfirmationDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ConfirmationDialog", {
    enumerable: true,
    get: function() {
        return ConfirmationDialog;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _renderer = require("../../features/renderer");
var _Delineator = require("../Delineator");
var _Window = require("../Window");
var ConfirmationDialog = function(param) {
    var snapName = param.snapName, snapId = param.snapId, node = param.node, onCancel = param.onCancel, onApprove = param.onApprove;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_Window.Window, {
        snapName: snapName,
        snapId: snapId,
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                margin: "4",
                marginTop: "0",
                flex: "1",
                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_Delineator.Delineator, {
                    snapName: snapName,
                    children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_renderer.Renderer, {
                        node: node
                    })
                })
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
                borderTop: "1px solid",
                borderTopColor: "border.default",
                paddingTop: "4",
                paddingX: "4",
                gap: "4",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Button, {
                        variant: "outline",
                        flex: "1",
                        onClick: onCancel,
                        children: "Cancel"
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Button, {
                        variant: "primary",
                        flex: "1",
                        onClick: onApprove,
                        children: "Approve"
                    })
                ]
            })
        ]
    });
};
