// @metamask/snaps-simulator/dist/cjs/features/handlers/cronjobs/slice.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    INITIAL_STATE: function() {
        return INITIAL_STATE;
    },
    cronjob: function() {
        return cronjob;
    },
    setCronjobRequest: function() {
        return setCronjobRequest;
    },
    setCronjobRequestFromHistory: function() {
        return setCronjobRequestFromHistory;
    },
    setCronjobResponse: function() {
        return setCronjobResponse;
    },
    clearCronjobResponse: function() {
        return clearCronjobResponse;
    },
    getCronjobRequest: function() {
        return getCronjobRequest;
    }
});
var _snapsutils = require("@metamask/snaps-utils");
var _toolkit = require("@reduxjs/toolkit");
var _slice = require("../slice");
var INITIAL_STATE = {
    request: {
        origin: ""
    },
    response: null,
    history: []
};
var slice = (0, _slice.createHandlerSlice)({
    name: _snapsutils.HandlerType.OnCronjob,
    initialState: INITIAL_STATE
});
var cronjob = slice.reducer;
var _slice_actions = slice.actions;
var setCronjobRequest = _slice_actions.setRequest, setCronjobRequestFromHistory = _slice_actions.setRequestFromHistory, setCronjobResponse = _slice_actions.setResponse, clearCronjobResponse = _slice_actions.clearResponse;
var getCronjobRequest = (0, _toolkit.createSelector)(function(state) {
    return state[_snapsutils.HandlerType.OnCronjob];
}, function(state) {
    return state.request;
});
