// @metamask/snaps-simulator/dist/cjs/store/reducer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "reducer", {
    enumerable: true,
    get: function() {
        return reducer;
    }
});
var _snapsutils = require("@metamask/snaps-utils");
var _toolkit = require("@reduxjs/toolkit");
var _slice = require("../features/configuration/slice");
var _slice1 = require("../features/console/slice");
var _slice2 = require("../features/handlers/cronjobs/slice");
var _slice3 = require("../features/handlers/json-rpc/slice");
var _slice4 = require("../features/handlers/transactions/slice");
var _slice5 = require("../features/manifest/slice");
var _slice6 = require("../features/notifications/slice");
var _slice7 = require("../features/simulation/slice");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
var _obj;
var reducer = (0, _toolkit.combineReducers)((_obj = {
    configuration: _slice.configuration,
    console: _slice1.console,
    manifest: _slice5.manifest,
    notifications: _slice6.notifications,
    simulation: _slice7.simulation
}, _define_property(_obj, _snapsutils.HandlerType.OnRpcRequest, _slice3.jsonRpc), _define_property(_obj, _snapsutils.HandlerType.OnCronjob, _slice2.cronjob), _define_property(_obj, _snapsutils.HandlerType.OnTransaction, _slice4.transactions), _obj));
