// @metamask/snaps-simulator/dist/esm/features/manifest/Manifest.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Flex, Tab, TabList, TabPanel, TabPanels, Tabs } from "@chakra-ui/react";
import { Editor } from "../../components";
import { useSelector } from "../../hooks";
import { getSnapManifest } from "../simulation";
import { Validation } from "./components";
/**
 * Manifest page, which displays the manifest of a snap, and any errors that
 * may be present.
 *
 * @returns The Manifest component.
 */ export var Manifest = function() {
    var manifest = useSelector(getSnapManifest);
    return /*#__PURE__*/ _jsx(Flex, {
        width: "100%",
        direction: "column",
        overflow: "hidden",
        children: /*#__PURE__*/ _jsxs(Flex, {
            direction: "row",
            flex: "1",
            overflow: "hidden",
            children: [
                /*#__PURE__*/ _jsx(Flex, {
                    direction: "column",
                    flex: "1",
                    width: "50%",
                    overflow: "hidden",
                    children: /*#__PURE__*/ _jsxs(Tabs, {
                        display: "flex",
                        flexDirection: "column",
                        flex: "1",
                        overflow: "hidden",
                        children: [
                            /*#__PURE__*/ _jsx(TabList, {
                                children: /*#__PURE__*/ _jsx(Tab, {
                                    children: "Validation"
                                })
                            }),
                            /*#__PURE__*/ _jsx(TabPanels, {
                                display: "flex",
                                flexDirection: "column",
                                flex: "1",
                                overflow: "hidden",
                                children: /*#__PURE__*/ _jsx(TabPanel, {
                                    display: "flex",
                                    flexDirection: "column",
                                    flex: "1",
                                    padding: "0",
                                    children: /*#__PURE__*/ _jsx(Validation, {})
                                })
                            })
                        ]
                    })
                }),
                /*#__PURE__*/ _jsx(Box, {
                    display: "flex",
                    flexDirection: "column",
                    flex: "1",
                    width: "50%",
                    borderLeft: "1px solid",
                    borderColor: "border.default",
                    children: /*#__PURE__*/ _jsxs(Tabs, {
                        display: "flex",
                        flexDirection: "column",
                        flex: "1",
                        overflow: "hidden",
                        children: [
                            /*#__PURE__*/ _jsx(TabList, {
                                children: /*#__PURE__*/ _jsx(Tab, {
                                    children: "Manifest"
                                })
                            }),
                            /*#__PURE__*/ _jsx(TabPanels, {
                                display: "flex",
                                flexDirection: "column",
                                flex: "1",
                                overflow: "hidden",
                                children: /*#__PURE__*/ _jsx(TabPanel, {
                                    display: "flex",
                                    flexDirection: "column",
                                    flex: "1",
                                    padding: "0",
                                    children: /*#__PURE__*/ _jsx(Editor, {
                                        border: "none",
                                        value: JSON.stringify(manifest, null, 2),
                                        options: {
                                            readOnly: true,
                                            wordWrap: "on"
                                        }
                                    })
                                })
                            })
                        ]
                    })
                })
            ]
        })
    });
};
