// @metamask/snaps-simulator/dist/cjs/routes.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "router", {
    enumerable: true,
    get: function() {
        return router;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _snapsutils = require("@metamask/snaps-utils");
var _reactrouterdom = require("react-router-dom");
var _features = require("./features");
var router = (0, _reactrouterdom.createHashRouter)((0, _reactrouterdom.createRoutesFromElements)(/*#__PURE__*/ (0, _jsxruntime.jsxs)(_reactrouterdom.Route, {
    element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Layout, {}),
    children: [
        /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Route, {
            path: "/",
            element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Navigate, {
                to: "/handler/".concat(_snapsutils.HandlerType.OnRpcRequest),
                replace: true
            })
        }),
        /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Route, {
            path: "/builder",
            element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Builder, {})
        }),
        /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Route, {
            path: "/manifest",
            element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Manifest, {})
        }),
        /*#__PURE__*/ (0, _jsxruntime.jsxs)(_reactrouterdom.Route, {
            path: "/handler",
            element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Handler, {}),
            children: [
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Route, {
                    path: "/handler/".concat(_snapsutils.HandlerType.OnRpcRequest),
                    element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.JsonRpc, {})
                }),
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Route, {
                    path: "/handler/".concat(_snapsutils.HandlerType.OnCronjob),
                    element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Cronjobs, {})
                }),
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Route, {
                    path: "/handler/".concat(_snapsutils.HandlerType.OnTransaction),
                    element: /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Transactions, {})
                })
            ]
        })
    ]
})));
