// @metamask/snaps-simulator/dist/cjs/utils/render.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "render", {
    enumerable: true,
    get: function() {
        return render;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@testing-library/react");
var _reactrouterdom = require("react-router-dom");
var _components = require("../components");
var _store = require("../store");
function render(component) {
    var route = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : "/";
    var store = (0, _store.createStore)();
    return (0, _react.render)(component, {
        wrapper: function(param) {
            var children = param.children;
            return /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Root, {
                store: store,
                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.MemoryRouter, {
                    initialEntries: [
                        route
                    ],
                    initialIndex: 0,
                    children: children
                })
            });
        }
    });
}
