// @metamask/snaps-simulator/dist/cjs/components/Author.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Author", {
    enumerable: true,
    get: function() {
        return Author;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _SnapIcon = require("./SnapIcon");
var Author = function(param) {
    var snapName = param.snapName, snapId = param.snapId;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        gap: "2",
        marginX: "4",
        marginY: "4",
        border: "1px solid",
        borderColor: "border.default",
        borderRadius: "32px",
        padding: "1",
        alignItems: "center",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_SnapIcon.SnapIcon, {
                snapName: snapName
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
                direction: "column",
                justify: "center",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                        fontSize: "sm",
                        lineHeight: "157%",
                        fontFamily: "custom",
                        children: snapName
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                        fontSize: "xs",
                        lineHeight: "167%",
                        fontFamily: "custom",
                        color: "text.alternative",
                        children: snapId
                    })
                ]
            })
        ]
    });
};
