// @metamask/snaps-simulator/dist/cjs/components/Delineator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Delineator", {
    enumerable: true,
    get: function() {
        return Delineator;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _Icon = require("./Icon");
var Delineator = function(param) {
    var snapName = param.snapName, children = param.children;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
        border: "1px solid",
        borderColor: "border.default",
        borderRadius: "md",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
                direction: "row",
                alignItems: "center",
                padding: "1.5",
                borderBottom: "1px solid",
                borderColor: "border.default",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_Icon.Icon, {
                        icon: "snap",
                        width: "16px",
                        marginRight: "1"
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Text, {
                        fontFamily: "custom",
                        fontSize: "xs",
                        children: [
                            "Content from ",
                            snapName
                        ]
                    })
                ]
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                padding: "3",
                children: children
            })
        ]
    });
};
