// @metamask/snaps-simulator/dist/cjs/components/Prefill.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Prefill", {
    enumerable: true,
    get: function() {
        return Prefill;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _react1 = require("react");
var _Icon = require("./Icon");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
function _object_without_properties(source, excluded) {
    if (source == null) return {};
    var target = _object_without_properties_loose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for(i = 0; i < sourceSymbolKeys.length; i++){
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) continue;
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
            target[key] = source[key];
        }
    }
    return target;
}
function _object_without_properties_loose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for(i = 0; i < sourceKeys.length; i++){
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        target[key] = source[key];
    }
    return target;
}
var Prefill = /*#__PURE__*/ (0, _react1.forwardRef)(function(_param, ref) /*#__PURE__*/ {
    var children = _param.children, _param_icon = _param.icon, icon = _param_icon === void 0 ? "darkArrowTopRightIcon" : _param_icon, _param_iconLocation = _param.iconLocation, iconLocation = _param_iconLocation === void 0 ? "right" : _param_iconLocation, props = _object_without_properties(_param, [
        "children",
        "icon",
        "iconLocation"
    ]);
    return (0, _jsxruntime.jsxs)(_react.Tag, _object_spread_props(_object_spread({
        ref: ref,
        width: "fit-content",
        fontWeight: "400",
        fontSize: "xs",
        fontFamily: "code",
        borderRadius: "xl",
        background: "background.alternative",
        _hover: {
            background: "info.muted"
        }
    }, props), {
        children: [
            iconLocation === "left" && /*#__PURE__*/ (0, _jsxruntime.jsx)(_Icon.Icon, {
                icon: icon,
                width: "10px",
                marginRight: "1"
            }),
            children,
            iconLocation === "right" && /*#__PURE__*/ (0, _jsxruntime.jsx)(_Icon.Icon, {
                icon: icon,
                width: "10px",
                marginLeft: "1"
            })
        ]
    }));
});
