// @metamask/snaps-simulator/dist/cjs/components/Root.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Root", {
    enumerable: true,
    get: function() {
        return Root;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _reactdndtreeview = require("@minoru/react-dnd-treeview");
var _react1 = require("react");
var _reactdnd = require("react-dnd");
var _reactredux = require("react-redux");
var _features = require("../features");
var _theme = require("../theme");
var Root = function(param) {
    var store = param.store, children = param.children;
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react1.StrictMode, {
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactredux.Provider, {
            store: store,
            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.ChakraProvider, {
                theme: _theme.theme,
                children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_reactdnd.DndProvider, {
                    backend: _reactdndtreeview.MultiBackend,
                    options: (0, _reactdndtreeview.getBackendOptions)(),
                    children: [
                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_features.Notifications, {}),
                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.ColorModeScript, {
                            initialColorMode: _theme.theme.config.initialColorMode
                        }),
                        children
                    ]
                })
            })
        })
    });
};
