// @metamask/snaps-simulator/dist/cjs/components/Window.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Window", {
    enumerable: true,
    get: function() {
        return Window;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _Author = require("./Author");
var Window = function(param) {
    var snapName = param.snapName, snapId = param.snapId, children = param.children, _param_showAuthorship = param.showAuthorship, showAuthorship = _param_showAuthorship === void 0 ? true : _param_showAuthorship;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        direction: "column",
        boxShadow: "lg",
        maxWidth: "360px",
        height: "600px",
        paddingY: "4",
        overflowY: "scroll",
        children: [
            showAuthorship && /*#__PURE__*/ (0, _jsxruntime.jsx)(_Author.Author, {
                snapName: snapName,
                snapId: snapId
            }),
            children
        ]
    });
};
