// @metamask/snaps-simulator/dist/cjs/features/handlers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    Cronjobs: function() {
        return _cronjobs.Cronjobs;
    },
    cronjob: function() {
        return _cronjobs.cronjob;
    },
    JsonRpc: function() {
        return _jsonrpc.JsonRpc;
    },
    jsonRpc: function() {
        return _jsonrpc.jsonRpc;
    },
    Transactions: function() {
        return _transactions.Transactions;
    },
    transactions: function() {
        return _transactions.transactions;
    }
});
_export_star(require("./components"), exports);
var _cronjobs = require("./cronjobs");
var _jsonrpc = require("./json-rpc");
var _transactions = require("./transactions");
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
