// @metamask/snaps-simulator/dist/cjs/features/handlers/slice.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "createHandlerSlice", {
    enumerable: true,
    get: function() {
        return createHandlerSlice;
    }
});
var _toolkit = require("@reduxjs/toolkit");
function createHandlerSlice(param) {
    var name = param.name, initialState = param.initialState;
    var slice = (0, _toolkit.createSlice)({
        name: name,
        initialState: initialState,
        reducers: {
            setRequest: function(state, action) {
                // `immer` does not work well with generic types, so we have to cast.
                state.request = action.payload;
                state.history.push({
                    date: new Date(),
                    request: action.payload
                });
                state.pending = true;
            },
            setRequestFromHistory: function(state, action) {
                // `immer` does not work well with generic types, so we have to cast.
                state.request = action.payload;
            },
            setResponse: function(state, action) {
                // `immer` does not work well with generic types, so we have to cast.
                state.response = action.payload;
                state.pending = false;
            },
            clearResponse: function(state) {
                state.response = null;
            }
        }
    });
    return slice;
}
