// @metamask/snaps-simulator/dist/esm/features/handlers/components/HistoryItem.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Flex, ListItem, Popover, PopoverContent, PopoverTrigger, Text, useTabsContext } from "@chakra-ui/react";
import { formatDistance } from "date-fns";
import { Icon } from "../../../components";
import { useDispatch, useHandler } from "../../../hooks";
export var HistoryItem = function(param) {
    var item = param.item;
    var _item_request_request, _item_request, _item_request1;
    var dispatch = useDispatch();
    var tab = useTabsContext();
    var handler = useHandler();
    var handleClick = function() {
        dispatch({
            type: "".concat(handler, "/setRequestFromHistory"),
            payload: item.request
        });
        tab.setSelectedIndex(0);
    };
    return /*#__PURE__*/ _jsx(ListItem, {
        onClick: handleClick,
        children: /*#__PURE__*/ _jsxs(Popover, {
            trigger: "hover",
            openDelay: 10,
            children: [
                /*#__PURE__*/ _jsx(PopoverTrigger, {
                    children: /*#__PURE__*/ _jsxs(Flex, {
                        margin: "2",
                        padding: "2",
                        borderRadius: "md",
                        justifyContent: "space-between",
                        alignItems: "center",
                        _hover: {
                            cursor: "pointer",
                            background: "background.hover"
                        },
                        gap: "2",
                        children: [
                            /*#__PURE__*/ _jsxs(Box, {
                                overflow: "hidden",
                                children: [
                                    /*#__PURE__*/ _jsx(Text, {
                                        fontSize: "sm",
                                        fontWeight: "600",
                                        overflow: "hidden",
                                        textOverflow: "ellipsis",
                                        whiteSpace: "nowrap",
                                        children: (_item_request = item.request) === null || _item_request === void 0 ? void 0 : (_item_request_request = _item_request.request) === null || _item_request_request === void 0 ? void 0 : _item_request_request.method
                                    }),
                                    /*#__PURE__*/ _jsx(Text, {
                                        color: "text.muted",
                                        fontSize: "sm",
                                        children: formatDistance(item.date, new Date(), {
                                            addSuffix: true
                                        })
                                    })
                                ]
                            }),
                            /*#__PURE__*/ _jsx(Box, {
                                children: /*#__PURE__*/ _jsx(Icon, {
                                    icon: "arrowTopRight",
                                    width: "10px"
                                })
                            })
                        ]
                    })
                }),
                /*#__PURE__*/ _jsx(PopoverContent, {
                    padding: "4",
                    fontFamily: "code",
                    fontSize: "xs",
                    overflow: "hidden",
                    textOverflow: "ellipsis",
                    whiteSpace: "nowrap",
                    children: /*#__PURE__*/ _jsx(Text, {
                        display: "block",
                        as: "pre",
                        overflow: "hidden",
                        textOverflow: "ellipsis",
                        whiteSpace: "pre",
                        children: JSON.stringify((_item_request1 = item.request) === null || _item_request1 === void 0 ? void 0 : _item_request1.request, null, 2)
                    })
                })
            ]
        })
    });
};
