// @metamask/snaps-simulator/dist/esm/features/handlers/components/Response.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Center, Heading, Text, Box, Skeleton } from "@chakra-ui/react";
import { HandlerType } from "@metamask/snaps-utils";
import { Delineator, Editor, Icon, Window } from "../../../components";
import { useSelector, useHandler } from "../../../hooks";
import { getSnapId } from "../../configuration";
import { Renderer } from "../../renderer";
import { getSnapName } from "../../simulation";
export var Response = function() {
    var _response_result;
    var handler = useHandler();
    var response = useSelector(function(state) {
        return state[handler].response;
    });
    var snapId = useSelector(getSnapId);
    var snapName = useSelector(getSnapName);
    if (!response) {
        return /*#__PURE__*/ _jsxs(Center, {
            background: "background.alternative",
            flex: "1",
            flexDirection: "column",
            children: [
                /*#__PURE__*/ _jsx(Icon, {
                    icon: "computer",
                    width: "34px",
                    height: "auto",
                    marginBottom: "1.5"
                }),
                /*#__PURE__*/ _jsx(Heading, {
                    as: "h3",
                    fontSize: "sm",
                    fontWeight: "700",
                    color: "text.muted",
                    marginBottom: "1",
                    children: "No response yet"
                }),
                /*#__PURE__*/ _jsxs(Text, {
                    fontSize: "xs",
                    textAlign: "center",
                    color: "text.muted",
                    children: [
                        "Create a request via the",
                        /*#__PURE__*/ _jsx("br", {}),
                        "left-side config to get started."
                    ]
                })
            ]
        });
    }
    // TODO: Fix this type cast.
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
    var content = (_response_result = response.result) === null || _response_result === void 0 ? void 0 : _response_result.content;
    if (handler === HandlerType.OnTransaction && content) {
        return /*#__PURE__*/ _jsx(Box, {
            margin: "4",
            children: /*#__PURE__*/ _jsx(Window, {
                snapName: snapName,
                snapId: snapId,
                showAuthorship: false,
                children: /*#__PURE__*/ _jsxs(Box, {
                    margin: "4",
                    flex: "1",
                    children: [
                        /*#__PURE__*/ _jsx(Skeleton, {
                            height: "38px",
                            mb: "4",
                            speed: 3
                        }),
                        /*#__PURE__*/ _jsx(Skeleton, {
                            height: "285px",
                            mb: "4",
                            speed: 3
                        }),
                        /*#__PURE__*/ _jsx(Delineator, {
                            snapName: snapName,
                            children: /*#__PURE__*/ _jsx(Renderer, {
                                node: content
                            })
                        })
                    ]
                })
            })
        });
    }
    return /*#__PURE__*/ _jsx(Editor, {
        border: "none",
        value: JSON.stringify(response, null, 2),
        options: {
            readOnly: true,
            wordWrap: "on"
        }
    });
};
