// @metamask/snaps-simulator/dist/esm/routes.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { HandlerType } from "@metamask/snaps-utils";
import { createHashRouter, createRoutesFromElements, Navigate, Route } from "react-router-dom";
import { Cronjobs, JsonRpc, Layout, Transactions, Handler, Manifest, Builder } from "./features";
export var router = createHashRouter(createRoutesFromElements(/*#__PURE__*/ _jsxs(Route, {
    element: /*#__PURE__*/ _jsx(Layout, {}),
    children: [
        /*#__PURE__*/ _jsx(Route, {
            path: "/",
            element: /*#__PURE__*/ _jsx(Navigate, {
                to: "/handler/".concat(HandlerType.OnRpcRequest),
                replace: true
            })
        }),
        /*#__PURE__*/ _jsx(Route, {
            path: "/builder",
            element: /*#__PURE__*/ _jsx(Builder, {})
        }),
        /*#__PURE__*/ _jsx(Route, {
            path: "/manifest",
            element: /*#__PURE__*/ _jsx(Manifest, {})
        }),
        /*#__PURE__*/ _jsxs(Route, {
            path: "/handler",
            element: /*#__PURE__*/ _jsx(Handler, {}),
            children: [
                /*#__PURE__*/ _jsx(Route, {
                    path: "/handler/".concat(HandlerType.OnRpcRequest),
                    element: /*#__PURE__*/ _jsx(JsonRpc, {})
                }),
                /*#__PURE__*/ _jsx(Route, {
                    path: "/handler/".concat(HandlerType.OnCronjob),
                    element: /*#__PURE__*/ _jsx(Cronjobs, {})
                }),
                /*#__PURE__*/ _jsx(Route, {
                    path: "/handler/".concat(HandlerType.OnTransaction),
                    element: /*#__PURE__*/ _jsx(Transactions, {})
                })
            ]
        })
    ]
})));
