// @metamask/snaps-simulator/dist/esm/features/renderer/components/Text.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Text as ChakraText } from "@chakra-ui/react";
import { isComponent } from "@metamask/snaps-ui";
import { assert } from "@metamask/utils";
import ReactMarkdown from "react-markdown";
export var Text = function(param) {
    var node = param.node, id = param.id;
    assert(isComponent(node), "Expected value to be a valid UI component.");
    assert(node.type === "text", "Expected value to be a text component.");
    return /*#__PURE__*/ _jsx(ReactMarkdown, {
        allowedElements: [
            "p",
            "strong",
            "em"
        ],
        components: {
            p: function(param) {
                var value = param.children;
                return /*#__PURE__*/ _jsx(ChakraText, {
                    fontFamily: "custom",
                    fontSize: "sm",
                    paddingBottom: "1",
                    children: value
                });
            }
        },
        children: node.value
    }, "".concat(id, "-text"));
};
