// @metamask/snaps-simulator/dist/esm/features/handlers/slice.js
import { createSlice } from "@reduxjs/toolkit";
/**
 * Create a slice for a handler.
 *
 * @param options - Options for the slice.
 * @param options.name - The name of the slice.
 * @param options.initialState - The initial state of the slice.
 * @returns The slice.
 */ export function createHandlerSlice(param) {
    var name = param.name, initialState = param.initialState;
    var slice = createSlice({
        name: name,
        initialState: initialState,
        reducers: {
            setRequest: function(state, action) {
                // `immer` does not work well with generic types, so we have to cast.
                state.request = action.payload;
                state.history.push({
                    date: new Date(),
                    request: action.payload
                });
                state.pending = true;
            },
            setRequestFromHistory: function(state, action) {
                // `immer` does not work well with generic types, so we have to cast.
                state.request = action.payload;
            },
            setResponse: function(state, action) {
                // `immer` does not work well with generic types, so we have to cast.
                state.response = action.payload;
                state.pending = false;
            },
            clearResponse: function(state) {
                state.response = null;
            }
        }
    });
    return slice;
}
