// @metamask/snaps-controllers/dist/cjs/snaps/endowments/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    endowmentPermissionBuilders: function() {
        return endowmentPermissionBuilders;
    },
    endowmentCaveatSpecifications: function() {
        return endowmentCaveatSpecifications;
    },
    endowmentCaveatMappers: function() {
        return endowmentCaveatMappers;
    },
    handlerEndowments: function() {
        return handlerEndowments;
    },
    getRpcCaveatOrigins: function() {
        return _rpc.getRpcCaveatOrigins;
    },
    getTransactionOriginCaveat: function() {
        return _transactioninsight.getTransactionOriginCaveat;
    }
});
const _snapsutils = require("@metamask/snaps-utils");
const _cronjob = require("./cronjob");
const _ethereumprovider = require("./ethereum-provider");
const _longrunning = require("./long-running");
const _networkaccess = require("./network-access");
const _rpc = require("./rpc");
const _transactioninsight = require("./transaction-insight");
const _webassembly = require("./web-assembly");
_export_star(require("./enum"), exports);
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
const endowmentPermissionBuilders = {
    [_networkaccess.networkAccessEndowmentBuilder.targetName]: _networkaccess.networkAccessEndowmentBuilder,
    [_longrunning.longRunningEndowmentBuilder.targetName]: _longrunning.longRunningEndowmentBuilder,
    [_transactioninsight.transactionInsightEndowmentBuilder.targetName]: _transactioninsight.transactionInsightEndowmentBuilder,
    [_cronjob.cronjobEndowmentBuilder.targetName]: _cronjob.cronjobEndowmentBuilder,
    [_ethereumprovider.ethereumProviderEndowmentBuilder.targetName]: _ethereumprovider.ethereumProviderEndowmentBuilder,
    [_rpc.rpcEndowmentBuilder.targetName]: _rpc.rpcEndowmentBuilder,
    [_webassembly.webAssemblyEndowmentBuilder.targetName]: _webassembly.webAssemblyEndowmentBuilder
};
const endowmentCaveatSpecifications = {
    ..._cronjob.cronjobCaveatSpecifications,
    ..._transactioninsight.transactionInsightCaveatSpecifications,
    ..._rpc.rpcCaveatSpecifications
};
const endowmentCaveatMappers = {
    [_cronjob.cronjobEndowmentBuilder.targetName]: _cronjob.getCronjobCaveatMapper,
    [_transactioninsight.transactionInsightEndowmentBuilder.targetName]: _transactioninsight.getTransactionInsightCaveatMapper,
    [_rpc.rpcEndowmentBuilder.targetName]: _rpc.getRpcCaveatMapper
};
const handlerEndowments = {
    [_snapsutils.HandlerType.OnRpcRequest]: _rpc.rpcEndowmentBuilder.targetName,
    [_snapsutils.HandlerType.OnTransaction]: _transactioninsight.transactionInsightEndowmentBuilder.targetName,
    [_snapsutils.HandlerType.OnCronjob]: _cronjob.cronjobEndowmentBuilder.targetName
};
