// @metamask/snaps-controllers/dist/cjs/snaps/endowments/long-running.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "longRunningEndowmentBuilder", {
    enumerable: true,
    get: function() {
        return longRunningEndowmentBuilder;
    }
});
const _permissioncontroller = require("@metamask/permission-controller");
const _enum = require("./enum");
const permissionName = _enum.SnapEndowments.LongRunning;
/**
 * `endowment:long-running` returns nothing; it is intended to be used as a flag
 * by the `SnapController` to make it ignore the request processing timeout
 * during snap lifecycle management. Essentially, it allows a snap to take an
 * infinite amount of time to process a request.
 *
 * @param _builderOptions - Optional specification builder options.
 * @returns The specification for the long-running endowment.
 */ const specificationBuilder = (_builderOptions)=>{
    return {
        permissionType: _permissioncontroller.PermissionType.Endowment,
        targetName: permissionName,
        allowedCaveats: null,
        endowmentGetter: (_getterOptions)=>undefined,
        subjectTypes: [
            _permissioncontroller.SubjectType.Snap
        ]
    };
};
const longRunningEndowmentBuilder = Object.freeze({
    targetName: permissionName,
    specificationBuilder
});
