// @metamask/snaps-simulator/dist/esm/features/manifest/components/Item.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Accordion, AccordionButton, AccordionIcon, AccordionItem, AccordionPanel, Box, Flex, ListItem, Tag, Text } from "@chakra-ui/react";
import { Icon } from "../../../components";
var BaseItem = function(param) {
    var isValid = param.isValid, name = param.name, manifestName = param.manifestName;
    return /*#__PURE__*/ _jsxs(Flex, {
        alignItems: "center",
        gap: "2",
        paddingY: "2",
        marginX: "4",
        marginY: "2",
        flex: "1",
        children: [
            /*#__PURE__*/ _jsx(Icon, {
                icon: isValid ? "playSuccess" : "errorTriangle",
                width: "14px"
            }),
            /*#__PURE__*/ _jsx(Text, {
                fontSize: "sm",
                fontWeight: "500",
                children: name
            }),
            /*#__PURE__*/ _jsx(Tag, {
                variant: "code",
                background: "background.alternative",
                color: "text.alternative",
                fontSize: "xs",
                children: manifestName
            })
        ]
    });
};
/**
 * A manifest validation item.
 *
 * @param props - The Item component props.
 * @param props.isValid - Whether the item is valid.
 * @param props.name - The name of the item.
 * @param props.manifestName - The name of the item in the manifest.
 * @param props.message - The validation message.
 * @returns The Item component.
 */ export var Item = function(param) {
    var isValid = param.isValid, name = param.name, manifestName = param.manifestName, message = param.message;
    if (isValid || !message) {
        return /*#__PURE__*/ _jsx(ListItem, {
            children: /*#__PURE__*/ _jsx(BaseItem, {
                isValid: isValid,
                name: name,
                manifestName: manifestName
            })
        });
    }
    return /*#__PURE__*/ _jsx(ListItem, {
        children: /*#__PURE__*/ _jsx(Accordion, {
            allowToggle: true,
            children: /*#__PURE__*/ _jsxs(AccordionItem, {
                border: "none",
                children: [
                    /*#__PURE__*/ _jsxs(AccordionButton, {
                        padding: "0",
                        _hover: {
                            background: "none"
                        },
                        children: [
                            /*#__PURE__*/ _jsx(BaseItem, {
                                isValid: isValid,
                                name: name,
                                manifestName: manifestName
                            }),
                            /*#__PURE__*/ _jsx(AccordionIcon, {
                                marginRight: "4",
                                color: "text.muted"
                            })
                        ]
                    }),
                    /*#__PURE__*/ _jsx(AccordionPanel, {
                        padding: "0",
                        children: /*#__PURE__*/ _jsx(Box, {
                            background: "error.muted",
                            color: "error.default",
                            borderRadius: "4px",
                            sx: {
                                p: {
                                    color: "error.default",
                                    fontSize: "xs"
                                }
                            },
                            marginX: "4",
                            marginBottom: "4",
                            padding: "2",
                            children: /*#__PURE__*/ _jsx(Text, {
                                children: message
                            })
                        })
                    })
                ]
            })
        })
    });
};
