// @metamask/snaps-simulator/dist/esm/features/navigation/components/Bottom.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, List, Text } from "@chakra-ui/react";
import { Icon } from "../../../components";
import { useDispatch } from "../../../hooks";
import { openConfigurationModal } from "../../configuration";
import { Item } from "./Item";
export var Bottom = function() {
    var dispatch = useDispatch();
    var handleOpenConfiguration = function() {
        dispatch(openConfigurationModal());
    };
    return /*#__PURE__*/ _jsxs(List, {
        borderTop: "1px solid",
        borderTopColor: "border.default",
        padding: "2",
        children: [
            /*#__PURE__*/ _jsxs(Item, {
                path: "https://github.com/MetaMask/snaps/tree/main/packages/snaps-simulator",
                isExternal: true,
                tag: "github",
                children: [
                    /*#__PURE__*/ _jsx(Icon, {
                        icon: "gitHub",
                        width: "24px"
                    }),
                    /*#__PURE__*/ _jsxs(Box, {
                        children: [
                            /*#__PURE__*/ _jsx(Text, {
                                children: /*#__PURE__*/ _jsx(Box, {
                                    as: "span",
                                    fontWeight: "600",
                                    children: "GitHub"
                                })
                            }),
                            /*#__PURE__*/ _jsx(Text, {
                                fontSize: "sm",
                                marginTop: "1",
                                children: "Report an issue or contribute to the project"
                            })
                        ]
                    })
                ]
            }),
            /*#__PURE__*/ _jsxs(Item, {
                path: "#",
                onClick: handleOpenConfiguration,
                tag: "configuration",
                children: [
                    /*#__PURE__*/ _jsx(Icon, {
                        icon: "configuration",
                        width: "24px"
                    }),
                    /*#__PURE__*/ _jsxs(Box, {
                        children: [
                            /*#__PURE__*/ _jsx(Text, {
                                children: /*#__PURE__*/ _jsx(Box, {
                                    as: "span",
                                    fontWeight: "600",
                                    children: "Settings"
                                })
                            }),
                            /*#__PURE__*/ _jsx(Text, {
                                fontSize: "sm",
                                marginTop: "1",
                                children: "Configure the simulation environment"
                            })
                        ]
                    })
                ]
            })
        ]
    });
};
