// @metamask/snaps-simulator/dist/esm/features/navigation/components/Item.js
import { jsx as _jsx } from "react/jsx-runtime";
import { ListItem, Stack } from "@chakra-ui/react";
import { useMatch } from "react-router-dom";
import { Link } from "../../../components";
export var Item = function(param) {
    var path = param.path, tag = param.tag, _param_isExternal = param.isExternal, isExternal = _param_isExternal === void 0 ? false : _param_isExternal, onClick = param.onClick, children = param.children;
    var active = useMatch(path);
    var isActive = Boolean(active);
    return /*#__PURE__*/ _jsx(ListItem, {
        children: /*#__PURE__*/ _jsx(Link, {
            padding: "2",
            variant: isActive ? "navigation-active" : "navigation-default",
            to: path,
            display: "block",
            isExternal: isExternal,
            onClick: onClick,
            "data-testid": "navigation-".concat(tag),
            _hover: {
                textDecoration: "none",
                opacity: 1,
                background: "background.hover"
            },
            children: /*#__PURE__*/ _jsx(Stack, {
                direction: "row",
                align: "center",
                children: children
            })
        })
    });
};
