// @metamask/snaps-simulator/dist/esm/features/navigation/components/ManifestStatusIndicator.js
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
import { jsx as _jsx } from "react/jsx-runtime";
import { Box } from "@chakra-ui/react";
import { useSelector } from "../../../hooks";
import { getManifestStatus, ManifestStatus } from "../../manifest/slice";
var _obj;
var MANIFEST_COLORS = (_obj = {}, _define_property(_obj, ManifestStatus.Valid, "success.default"), _define_property(_obj, ManifestStatus.Invalid, "error.default"), _obj);
export var ManifestStatusIndicator = function() {
    var manifestStatus = useSelector(getManifestStatus);
    if (manifestStatus === ManifestStatus.Unknown) {
        return null;
    }
    return /*#__PURE__*/ _jsx(Box, {
        position: "absolute",
        bottom: "0",
        right: "0",
        width: "10px",
        height: "10px",
        background: MANIFEST_COLORS[manifestStatus],
        borderRadius: "5px"
    });
};
