// @metamask/snaps-jest/dist/cjs/options.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getOptions", {
    enumerable: true,
    get: function() {
        return getOptions;
    }
});
const _superstruct = require("superstruct");
const SnapsEnvironmentOptionsStruct = (0, _superstruct.type)({
    executionEnvironmentUrl: (0, _superstruct.optional)((0, _superstruct.string)()),
    simulatorUrl: (0, _superstruct.optional)((0, _superstruct.string)()),
    keepAlive: (0, _superstruct.defaulted)((0, _superstruct.boolean)(), false),
    server: (0, _superstruct.defaulted)((0, _superstruct.object)({
        enabled: (0, _superstruct.defaulted)((0, _superstruct.boolean)(), true),
        port: (0, _superstruct.defaulted)((0, _superstruct.number)(), 0),
        root: (0, _superstruct.defaulted)((0, _superstruct.string)(), process.cwd())
    }), {}),
    browser: (0, _superstruct.defaulted)((0, _superstruct.object)({
        headless: (0, _superstruct.defaulted)((0, _superstruct.boolean)(), true)
    }), {})
});
function getOptions(testEnvironmentOptions) {
    return (0, _superstruct.create)(testEnvironmentOptions, SnapsEnvironmentOptionsStruct);
}
