// @metamask/snaps-simulator/dist/esm/features/layout/components/Header.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Container, Stack } from "@chakra-ui/react";
import { Logo } from "../../../components";
import { Configuration } from "../../configuration";
import { StatusIndicator } from "../../status";
export var Header = function() {
    return /*#__PURE__*/ _jsxs(Container, {
        as: "header",
        size: "fullWidth",
        borderBottom: "muted",
        display: "flex",
        flexDirection: "row",
        justifyContent: "space-between",
        children: [
            /*#__PURE__*/ _jsx(Stack, {
                direction: "row",
                height: "7",
                align: "center",
                children: /*#__PURE__*/ _jsx(Logo, {})
            }),
            /*#__PURE__*/ _jsxs(Stack, {
                direction: "row",
                height: "7",
                align: "center",
                children: [
                    /*#__PURE__*/ _jsx(StatusIndicator, {}),
                    /*#__PURE__*/ _jsx(Configuration, {})
                ]
            })
        ]
    });
};
