// @metamask/snaps-simulator/dist/cjs/features/notifications/slice.test.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _rpcmethods = require("@metamask/rpc-methods");
var _slice = require("./slice");
describe("notifications slice", function() {
    describe("addNotification", function() {
        it("adds a notification to the state", function() {
            var result = (0, _slice.notifications)(_slice.INITIAL_NOTIFICATIONS_STATE, (0, _slice.addNotification)({
                id: "foo",
                message: "Hello, world!"
            }));
            expect(result.notifications).toHaveLength(1);
            expect(result.notifications).toStrictEqual([
                {
                    id: expect.any(String),
                    message: "Hello, world!",
                    type: _rpcmethods.NotificationType.InApp
                }
            ]);
        });
    });
    describe("addNativeNotification", function() {
        it("adds a native notification to the state", function() {
            var result = (0, _slice.notifications)(_slice.INITIAL_NOTIFICATIONS_STATE, (0, _slice.addNativeNotification)({
                id: "foo",
                message: "Hello, world!"
            }));
            expect(result.notifications).toHaveLength(0);
            expect(result.allNotifications).toHaveLength(1);
            expect(result.allNotifications).toStrictEqual([
                {
                    id: expect.any(String),
                    message: "Hello, world!",
                    type: _rpcmethods.NotificationType.Native
                }
            ]);
        });
    });
    describe("removeNotification", function() {
        it("removes a notification from the state", function() {
            var result = (0, _slice.notifications)({
                allNotifications: [
                    {
                        id: "1",
                        message: "Hello, world!",
                        type: _rpcmethods.NotificationType.Native
                    }
                ],
                notifications: [
                    {
                        id: "1",
                        message: "Hello, world!",
                        type: _rpcmethods.NotificationType.InApp
                    }
                ]
            }, (0, _slice.removeNotification)("1"));
            expect(result.notifications).toStrictEqual([]);
            expect(result.allNotifications).toStrictEqual([
                {
                    id: "1",
                    message: "Hello, world!",
                    type: _rpcmethods.NotificationType.Native
                }
            ]);
        });
    });
});
