// @metamask/snaps-simulator/dist/cjs/features/handlers/json-rpc/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    JSON_RPC_SCHEMA_URL: function() {
        return JSON_RPC_SCHEMA_URL;
    },
    JSON_RPC_SCHEMA: function() {
        return JSON_RPC_SCHEMA;
    },
    SAMPLE_JSON_RPC_REQUEST: function() {
        return SAMPLE_JSON_RPC_REQUEST;
    }
});
var JSON_RPC_SCHEMA_URL = "http://json-schema.org/draft-04/schema#";
var JSON_RPC_SCHEMA = {
    $schema: "http://json-schema.org/draft-04/schema#",
    type: "object",
    properties: {
        jsonrpc: {
            const: "2.0"
        },
        id: {
            oneOf: [
                {
                    type: "string"
                },
                {
                    type: "number"
                },
                {
                    type: "null"
                }
            ]
        },
        method: {
            type: "string"
        },
        params: {
            type: [
                "number",
                "string",
                "boolean",
                "object",
                "array",
                "null"
            ]
        }
    },
    required: [
        "jsonrpc",
        "id",
        "method"
    ],
    additionalProperties: false
};
var SAMPLE_JSON_RPC_REQUEST = JSON.stringify({
    jsonrpc: "2.0",
    id: 1,
    method: "subtract",
    params: [
        42,
        23
    ]
}, null, 2);
