// @metamask/snaps-simulator/dist/esm/features/navigation/Navigation.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Container, List, Stack, Text } from "@chakra-ui/react";
import { Icon } from "../../components";
import { useSelector } from "../../hooks";
import { Item, ManifestStatusIndicator } from "./components";
import { NavigationTag } from "./components/NavigationTag";
import { NAVIGATION_ITEMS } from "./items";
/**
 * The navigation component, which holds the navigation buttons.
 *
 * @returns The navigation component.
 */ export var Navigation = function() {
    var applicationState = useSelector(function(state) {
        return state;
    });
    return /*#__PURE__*/ _jsx(Container, {
        as: "nav",
        size: "fullWidth",
        display: "flex",
        flexDirection: "column",
        padding: "2",
        flex: "1",
        children: /*#__PURE__*/ _jsxs(Stack, {
            as: List,
            spacing: "2",
            flex: "1",
            children: [
                NAVIGATION_ITEMS.map(function(param) {
                    var condition = param.condition, icon = param.icon, label = param.label, tag = param.tag, description = param.description, path = param.path;
                    if (condition && !condition(applicationState)) {
                        return null;
                    }
                    return /*#__PURE__*/ _jsxs(Item, {
                        path: path,
                        tag: tag,
                        children: [
                            /*#__PURE__*/ _jsx(Icon, {
                                icon: icon
                            }),
                            /*#__PURE__*/ _jsxs(Box, {
                                children: [
                                    /*#__PURE__*/ _jsxs(Text, {
                                        children: [
                                            /*#__PURE__*/ _jsx(Box, {
                                                as: "span",
                                                fontWeight: "600",
                                                children: label
                                            }),
                                            " ",
                                            /*#__PURE__*/ _jsx(NavigationTag, {
                                                path: path,
                                                children: tag
                                            })
                                        ]
                                    }),
                                    /*#__PURE__*/ _jsx(Text, {
                                        fontSize: "sm",
                                        marginTop: "1",
                                        children: description
                                    })
                                ]
                            })
                        ]
                    }, path);
                }),
                /*#__PURE__*/ _jsxs(Item, {
                    tag: "manifest",
                    path: "/manifest",
                    children: [
                        /*#__PURE__*/ _jsxs(Box, {
                            position: "relative",
                            children: [
                                /*#__PURE__*/ _jsx(Icon, {
                                    icon: "manifest"
                                }),
                                /*#__PURE__*/ _jsx(ManifestStatusIndicator, {})
                            ]
                        }),
                        /*#__PURE__*/ _jsxs(Box, {
                            children: [
                                /*#__PURE__*/ _jsxs(Text, {
                                    children: [
                                        /*#__PURE__*/ _jsx(Box, {
                                            as: "span",
                                            fontWeight: "600",
                                            children: "Manifest"
                                        }),
                                        " ",
                                        /*#__PURE__*/ _jsx(NavigationTag, {
                                            path: "/manifest",
                                            children: "snap.manifest.json"
                                        })
                                    ]
                                }),
                                /*#__PURE__*/ _jsx(Text, {
                                    fontSize: "sm",
                                    marginTop: "1",
                                    children: "Validate the snap manifest"
                                })
                            ]
                        })
                    ]
                }, "manifest")
            ]
        })
    });
};
