// @metamask/snaps-utils/dist/cjs/handlers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    HandlerType: function() {
        return HandlerType;
    },
    SNAP_EXPORTS: function() {
        return SNAP_EXPORTS;
    }
});
var HandlerType;
(function(HandlerType) {
    HandlerType["OnRpcRequest"] = 'onRpcRequest';
    HandlerType["OnTransaction"] = 'onTransaction';
    HandlerType["OnCronjob"] = 'onCronjob';
    HandlerType["OnInstall"] = 'onInstall';
    HandlerType["OnUpdate"] = 'onUpdate';
})(HandlerType || (HandlerType = {}));
const SNAP_EXPORTS = {
    [HandlerType.OnRpcRequest]: {
        type: HandlerType.OnRpcRequest,
        required: true,
        validator: (snapExport)=>{
            return typeof snapExport === 'function';
        }
    },
    [HandlerType.OnTransaction]: {
        type: HandlerType.OnTransaction,
        required: true,
        validator: (snapExport)=>{
            return typeof snapExport === 'function';
        }
    },
    [HandlerType.OnCronjob]: {
        type: HandlerType.OnCronjob,
        required: true,
        validator: (snapExport)=>{
            return typeof snapExport === 'function';
        }
    },
    [HandlerType.OnInstall]: {
        type: HandlerType.OnInstall,
        required: false,
        validator: (snapExport)=>{
            return typeof snapExport === 'function';
        }
    },
    [HandlerType.OnUpdate]: {
        type: HandlerType.OnUpdate,
        required: false,
        validator: (snapExport)=>{
            return typeof snapExport === 'function';
        }
    }
};
