// @metamask/snaps-simulator/dist/cjs/features/builder/components/BaseNode.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BaseNode", {
    enumerable: true,
    get: function() {
        return BaseNode;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _utils = require("@metamask/utils");
var _components = require("../../../components");
var BaseNode = function(param) {
    var node = param.node, isDragging = param.isDragging, children = param.children, onClose = param.onClose;
    var _node_data;
    (0, _utils.assert)((_node_data = node.data) === null || _node_data === void 0 ? void 0 : _node_data.type, "Node must have a type.");
    var handleClick = function() {
        var _onClose;
        (_onClose = onClose) === null || _onClose === void 0 ? void 0 : _onClose(node);
    };
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        paddingY: "2",
        paddingLeft: "4",
        paddingRight: "2",
        background: "background.alternative",
        borderRadius: "lg",
        gap: "4",
        alignItems: "center",
        border: "1px solid",
        borderColor: "border.default",
        display: isDragging ? "none" : "flex",
        marginX: "4",
        cursor: node.id > 1 ? "move" : "default",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                icon: node.data.type,
                width: "16px"
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                fontWeight: "500",
                fontSize: "sm",
                textTransform: "capitalize",
                color: "text.muted",
                display: "inline-block",
                minWidth: "65px",
                children: node.data.type
            }),
            children,
            node.id >= 2 && /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                icon: "cross",
                width: "11px",
                marginX: "0.5",
                marginLeft: "auto",
                cursor: "pointer",
                onClick: handleClick
            })
        ]
    });
};
