// @metamask/snaps-simulator/dist/cjs/features/builder/components/EditableNode.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "EditableNode", {
    enumerable: true,
    get: function() {
        return EditableNode;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _utils = require("@metamask/utils");
var _react1 = require("react");
var _utils1 = require("../utils");
var _BaseNode = require("./BaseNode");
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
var EditableNode = function(param) {
    var node = param.node, depth = param.depth, isDragging = param.isDragging, onChange = param.onChange, onClose = param.onClose;
    var text = (0, _utils1.getNodeText)(node);
    (0, _utils.assert)(text !== null, "Node must have text.");
    var _useState = _sliced_to_array((0, _react1.useState)(text), 2), value = _useState[0], setValue = _useState[1];
    var handleChange = function(newValue) {
        var _onChange;
        setValue(newValue.target.value);
        (_onChange = onChange) === null || _onChange === void 0 ? void 0 : _onChange(node, newValue.target.value);
    };
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
        marginLeft: "".concat(depth * 16, "px"),
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_BaseNode.BaseNode, {
            node: node,
            isDragging: isDragging,
            onClose: onClose,
            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Input, {
                value: value,
                onChange: handleChange,
                fontSize: "sm",
                bg: "chakra-body-bg",
                borderColor: "border.default",
                outline: "none",
                _active: {
                    borderColor: "border.active",
                    outline: "none",
                    boxShadow: "none"
                },
                _focusVisible: {
                    borderColor: "border.active",
                    outline: "none",
                    boxShadow: "none"
                }
            })
        })
    });
};
