// @metamask/snaps-simulator/dist/cjs/features/builder/components/Node.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    EDITABLE_NODES: function() {
        return EDITABLE_NODES;
    },
    Node: function() {
        return Node;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _snapsui = require("@metamask/snaps-ui");
var _utils = require("@metamask/utils");
var _BaseNode = require("./BaseNode");
var _EditableNode = require("./EditableNode");
var EDITABLE_NODES = [
    _snapsui.NodeType.Heading,
    _snapsui.NodeType.Text,
    _snapsui.NodeType.Copyable
];
var Node = function(param) {
    var node = param.node, depth = param.depth, isDragging = param.isDragging, onChange = param.onChange, onClose = param.onClose;
    var _node_data;
    (0, _utils.assert)((_node_data = node.data) === null || _node_data === void 0 ? void 0 : _node_data.type, "Node must have a type.");
    if (EDITABLE_NODES.includes(node.data.type)) {
        return /*#__PURE__*/ (0, _jsxruntime.jsx)(_EditableNode.EditableNode, {
            node: node,
            depth: depth,
            isDragging: isDragging,
            onChange: onChange,
            onClose: onClose
        });
    }
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
        marginLeft: "".concat(depth * 16, "px"),
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_BaseNode.BaseNode, {
            node: node,
            isDragging: isDragging,
            onClose: onClose
        })
    });
};
