// @metamask/snaps-simulator/dist/cjs/features/builder/components/NodeRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "NodeRenderer", {
    enumerable: true,
    get: function() {
        return NodeRenderer;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _react1 = require("react");
var _reactredux = require("react-redux");
var _components = require("../../../components");
var _configuration = require("../../configuration");
var _renderer = require("../../renderer");
var _simulation = require("../../simulation");
var _utils = require("../utils");
var NodeRenderer = function(param) {
    var items = param.items;
    var snapId = (0, _reactredux.useSelector)(_configuration.getSnapId);
    var _useSelector;
    var snapName = (_useSelector = (0, _reactredux.useSelector)(_simulation.getSnapName)) !== null && _useSelector !== void 0 ? _useSelector : "Unknown";
    var node = (0, _react1.useMemo)(function() {
        return (0, _utils.nodeModelsToComponent)(items);
    }, [
        items
    ]);
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Window, {
        snapName: snapName,
        snapId: snapId,
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
            margin: "4",
            marginTop: "0",
            flex: "1",
            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Delineator, {
                snapName: snapName,
                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_renderer.Renderer, {
                    node: node
                })
            })
        })
    });
};
