// @metamask/snaps-simulator/dist/cjs/features/builder/components/Start.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Start", {
    enumerable: true,
    get: function() {
        return Start;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _components = require("../../../components");
var Start = function() {
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        flexDirection: "column",
        alignItems: "center",
        borderColor: "border.default",
        padding: "4",
        paddingTop: "12",
        borderRadius: "lg",
        background: "background.alternative",
        flex: "1",
        marginY: "-4",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                icon: "drag",
                marginBottom: "1.5"
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Heading, {
                as: "h3",
                fontSize: "sm",
                fontWeight: "700",
                color: "text.muted",
                marginBottom: "1",
                children: "Drag components in here"
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Text, {
                fontSize: "xs",
                fontWeight: "500",
                color: "text.muted",
                textAlign: "center",
                children: [
                    "Build your UI piece-by-piece using the",
                    /*#__PURE__*/ (0, _jsxruntime.jsx)("br", {}),
                    "predefined components above."
                ]
            })
        ]
    });
};
