// @metamask/snaps-simulator/dist/cjs/features/console/ConsoleContent.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ConsoleContent", {
    enumerable: true,
    get: function() {
        return ConsoleContent;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _hooks = require("../../hooks");
var _slice = require("./slice");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
var ConsoleContent = function() {
    var entries = (0, _hooks.useSelector)(_slice.getConsoleEntries);
    var _obj;
    var colors = (_obj = {}, _define_property(_obj, _slice.ConsoleEntryType.Log, "text.console"), _define_property(_obj, _slice.ConsoleEntryType.Error, "text.error"), _obj);
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_jsxruntime.Fragment, {
        children: entries.map(function(entry) {
            return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                textColor: colors[entry.type],
                fontFamily: "code",
                fontSize: "xs",
                whiteSpace: "pre-wrap",
                children: entry.message
            }, entry.date);
        })
    });
};
