// @metamask/snaps-simulator/dist/cjs/features/console/slice.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    ConsoleEntryType: function() {
        return ConsoleEntryType;
    },
    addDefault: function() {
        return addDefault;
    },
    addError: function() {
        return addError;
    },
    console: function() {
        return console;
    },
    getConsoleEntries: function() {
        return getConsoleEntries;
    }
});
var _toolkit = require("@reduxjs/toolkit");
var ConsoleEntryType;
(function(ConsoleEntryType) {
    ConsoleEntryType["Log"] = "log";
    ConsoleEntryType["Error"] = "error";
})(ConsoleEntryType || (ConsoleEntryType = {}));
var INITIAL_STATE = {
    entries: []
};
var slice = (0, _toolkit.createSlice)({
    name: "console",
    initialState: INITIAL_STATE,
    reducers: {
        addDefault: function addDefault(state, action) {
            state.entries.push({
                date: Date.now(),
                type: ConsoleEntryType.Log,
                message: action.payload
            });
        },
        addError: function addError(state, action) {
            var _action_payload_stack;
            state.entries.push({
                date: Date.now(),
                type: ConsoleEntryType.Error,
                message: (_action_payload_stack = action.payload.stack) !== null && _action_payload_stack !== void 0 ? _action_payload_stack : action.payload.message
            });
        }
    }
});
var _slice_actions = slice.actions;
var addDefault = _slice_actions.addDefault, addError = _slice_actions.addError;
var console = slice.reducer;
var getConsoleEntries = (0, _toolkit.createSelector)(function(state) {
    return state.console;
}, function(state) {
    return state.entries;
});
