// @metamask/snaps-simulator/dist/cjs/features/renderer/components/Text.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Text", {
    enumerable: true,
    get: function() {
        return Text;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _snapsui = require("@metamask/snaps-ui");
var _utils = require("@metamask/utils");
var _reactmarkdown = /*#__PURE__*/ _interop_require_default(require("react-markdown"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var Text = function(param) {
    var node = param.node, id = param.id;
    (0, _utils.assert)((0, _snapsui.isComponent)(node), "Expected value to be a valid UI component.");
    (0, _utils.assert)(node.type === "text", "Expected value to be a text component.");
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactmarkdown.default, {
        allowedElements: [
            "p",
            "strong",
            "em"
        ],
        components: {
            p: function(param) {
                var value = param.children;
                return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                    fontFamily: "custom",
                    fontSize: "sm",
                    paddingBottom: "1",
                    children: value
                });
            }
        },
        children: node.value
    }, "".concat(id, "-text"));
};
