// @metamask/snaps-simulator/dist/cjs/features/handlers/cronjobs/components/CronjobPrefills.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "CronjobPrefills", {
    enumerable: true,
    get: function() {
        return CronjobPrefills;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _hooks = require("../../../../hooks");
var _simulation = require("../../../simulation");
var _CronjobPrefill = require("./CronjobPrefill");
var CronjobPrefills = function(param) {
    var onClick = param.onClick;
    var _manifest_initialPermissions_endowmentcronjob, _manifest_initialPermissions, _manifest, _jobs;
    var manifest = (0, _hooks.useSelector)(_simulation.getSnapManifest);
    var jobs = (_manifest = manifest) === null || _manifest === void 0 ? void 0 : (_manifest_initialPermissions = _manifest.initialPermissions) === null || _manifest_initialPermissions === void 0 ? void 0 : (_manifest_initialPermissions_endowmentcronjob = _manifest_initialPermissions["endowment:cronjob"]) === null || _manifest_initialPermissions_endowmentcronjob === void 0 ? void 0 : _manifest_initialPermissions_endowmentcronjob.jobs;
    if (!((_jobs = jobs) === null || _jobs === void 0 ? void 0 : _jobs.length)) {
        return null;
    }
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
        marginBottom: "4",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                fontWeight: "500",
                fontSize: "xs",
                marginBottom: "1",
                children: "Manifest cronjobs"
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Flex, {
                gap: "2",
                children: jobs.map(function(param, index) {
                    var _param_request = param.request, method = _param_request.method, params = _param_request.params;
                    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_CronjobPrefill.CronjobPrefill, {
                        method: method,
                        params: params,
                        onClick: onClick
                    }, "cronjob-prefill-".concat(index));
                })
            })
        ]
    });
};
