// @metamask/snaps-simulator/dist/cjs/features/handlers/cronjobs/components/Request.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Request", {
    enumerable: true,
    get: function() {
        return Request;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _snapsutils = require("@metamask/snaps-utils");
var _reacthookform = require("react-hook-form");
var _components = require("../../../../components");
var _hooks = require("../../../../hooks");
var _simulation = require("../../../simulation");
var _schema = require("../../json-rpc/schema");
var _slice = require("../slice");
var _CronjobPrefills = require("./CronjobPrefills");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
var Request = function() {
    var _errors_origin;
    var _useSelector = (0, _hooks.useSelector)(_slice.getCronjobRequest), request = _useSelector.request, origin = _useSelector.origin;
    var _useForm = (0, _reacthookform.useForm)({
        defaultValues: {
            origin: origin !== null && origin !== void 0 ? origin : "",
            request: request ? JSON.stringify(request, null, 2) : _schema.SAMPLE_JSON_RPC_REQUEST
        }
    }), handleSubmit = _useForm.handleSubmit, register = _useForm.register, control = _useForm.control, setValue = _useForm.setValue, errors = _useForm.formState.errors;
    var dispatch = (0, _hooks.useDispatch)();
    var onSubmit = function(data) {
        dispatch((0, _simulation.sendRequest)({
            origin: data.origin,
            handler: _snapsutils.HandlerType.OnCronjob,
            request: JSON.parse(data.request)
        }));
    };
    var handlePrefill = function(data) {
        setValue("request", JSON.stringify(_object_spread({
            jsonrpc: "2.0",
            id: 1
        }, data), null, 2));
    };
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        as: "form",
        flexDirection: "column",
        flex: "1",
        /* eslint-disable-next-line @typescript-eslint/no-misused-promises */ onSubmit: handleSubmit(onSubmit),
        id: "request-form",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_CronjobPrefills.CronjobPrefills, {
                onClick: handlePrefill
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.FormControl, {
                isInvalid: Boolean(errors.origin),
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.FormLabel, {
                        htmlFor: "origin",
                        children: "Origin"
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Input, _object_spread({
                        id: "origin",
                        placeholder: "metamask.io",
                        fontFamily: "code"
                    }, register("origin"))),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.FormErrorMessage, {
                        children: (_errors_origin = errors.origin) === null || _errors_origin === void 0 ? void 0 : _errors_origin.message
                    })
                ]
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.FormControl, {
                isInvalid: Boolean(errors.request),
                display: "flex",
                flexDirection: "column",
                flex: "1",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.FormLabel, {
                        htmlFor: "request",
                        children: "Request"
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_reacthookform.Controller, {
                        control: control,
                        name: "request",
                        render: function(param) {
                            var _param_field = param.field, onChange = _param_field.onChange, value = _param_field.value;
                            return /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Editor, {
                                onChange: onChange,
                                value: value
                            });
                        }
                    })
                ]
            })
        ]
    });
};
