// @metamask/snaps-simulator/dist/cjs/features/manifest/components/Item.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Item", {
    enumerable: true,
    get: function() {
        return Item;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _components = require("../../../components");
var BaseItem = function(param) {
    var isValid = param.isValid, name = param.name, manifestName = param.manifestName;
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        alignItems: "center",
        gap: "2",
        paddingY: "2",
        marginX: "4",
        marginY: "2",
        flex: "1",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                icon: isValid ? "playSuccess" : "errorTriangle",
                width: "14px"
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                fontSize: "sm",
                fontWeight: "500",
                children: name
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Tag, {
                variant: "code",
                background: "background.alternative",
                color: "text.alternative",
                fontSize: "xs",
                children: manifestName
            })
        ]
    });
};
var Item = function(param) {
    var isValid = param.isValid, name = param.name, manifestName = param.manifestName, message = param.message;
    if (isValid || !message) {
        return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.ListItem, {
            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(BaseItem, {
                isValid: isValid,
                name: name,
                manifestName: manifestName
            })
        });
    }
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.ListItem, {
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Accordion, {
            allowToggle: true,
            children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.AccordionItem, {
                border: "none",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.AccordionButton, {
                        padding: "0",
                        _hover: {
                            background: "none"
                        },
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(BaseItem, {
                                isValid: isValid,
                                name: name,
                                manifestName: manifestName
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.AccordionIcon, {
                                marginRight: "4",
                                color: "text.muted"
                            })
                        ]
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.AccordionPanel, {
                        padding: "0",
                        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                            background: "error.muted",
                            color: "error.default",
                            borderRadius: "4px",
                            sx: {
                                p: {
                                    color: "error.default",
                                    fontSize: "xs"
                                }
                            },
                            marginX: "4",
                            marginBottom: "4",
                            padding: "2",
                            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                children: message
                            })
                        })
                    })
                ]
            })
        })
    });
};
