// @metamask/snaps-controllers/dist/esm/snaps/endowments/long-running.js
import { PermissionType, SubjectType } from '@metamask/permission-controller';
import { SnapEndowments } from './enum';
const permissionName = SnapEndowments.LongRunning;
/**
 * `endowment:long-running` returns nothing; it is intended to be used as a flag
 * by the `SnapController` to make it ignore the request processing timeout
 * during snap lifecycle management. Essentially, it allows a snap to take an
 * infinite amount of time to process a request.
 *
 * @param _builderOptions - Optional specification builder options.
 * @returns The specification for the long-running endowment.
 */ const specificationBuilder = (_builderOptions)=>{
    return {
        permissionType: PermissionType.Endowment,
        targetName: permissionName,
        allowedCaveats: null,
        endowmentGetter: (_getterOptions)=>undefined,
        subjectTypes: [
            SubjectType.Snap
        ]
    };
};
export const longRunningEndowmentBuilder = Object.freeze({
    targetName: permissionName,
    specificationBuilder
});
