// @metamask/snaps-simulator/dist/cjs/features/navigation/components/Bottom.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Bottom", {
    enumerable: true,
    get: function() {
        return Bottom;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _components = require("../../../components");
var _hooks = require("../../../hooks");
var _configuration = require("../../configuration");
var _Item = require("./Item");
var Bottom = function() {
    var dispatch = (0, _hooks.useDispatch)();
    var handleOpenConfiguration = function() {
        dispatch((0, _configuration.openConfigurationModal)());
    };
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.List, {
        borderTop: "1px solid",
        borderTopColor: "border.default",
        padding: "2",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_Item.Item, {
                path: "https://github.com/MetaMask/snaps/tree/main/packages/snaps-simulator",
                isExternal: true,
                tag: "github",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                        icon: "gitHub",
                        width: "24px"
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                                    as: "span",
                                    fontWeight: "600",
                                    children: "GitHub"
                                })
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                fontSize: "sm",
                                marginTop: "1",
                                children: "Report an issue or contribute to the project"
                            })
                        ]
                    })
                ]
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_Item.Item, {
                path: "#",
                onClick: handleOpenConfiguration,
                tag: "configuration",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                        icon: "configuration",
                        width: "24px"
                    }),
                    /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                                    as: "span",
                                    fontWeight: "600",
                                    children: "Settings"
                                })
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                fontSize: "sm",
                                marginTop: "1",
                                children: "Configure the simulation environment"
                            })
                        ]
                    })
                ]
            })
        ]
    });
};
