// @metamask/snaps-simulator/dist/cjs/features/navigation/components/Item.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Item", {
    enumerable: true,
    get: function() {
        return Item;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _reactrouterdom = require("react-router-dom");
var _components = require("../../../components");
var Item = function(param) {
    var path = param.path, tag = param.tag, _param_isExternal = param.isExternal, isExternal = _param_isExternal === void 0 ? false : _param_isExternal, onClick = param.onClick, children = param.children;
    var active = (0, _reactrouterdom.useMatch)(path);
    var isActive = Boolean(active);
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.ListItem, {
        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Link, {
            padding: "2",
            variant: isActive ? "navigation-active" : "navigation-default",
            to: path,
            display: "block",
            isExternal: isExternal,
            onClick: onClick,
            "data-testid": "navigation-".concat(tag),
            _hover: {
                textDecoration: "none",
                opacity: 1,
                background: "background.hover"
            },
            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Stack, {
                direction: "row",
                align: "center",
                children: children
            })
        })
    });
};
