// @metamask/snaps-simulator/dist/cjs/features/navigation/components/ManifestStatusIndicator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ManifestStatusIndicator", {
    enumerable: true,
    get: function() {
        return ManifestStatusIndicator;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _hooks = require("../../../hooks");
var _slice = require("../../manifest/slice");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
var _obj;
var MANIFEST_COLORS = (_obj = {}, _define_property(_obj, _slice.ManifestStatus.Valid, "success.default"), _define_property(_obj, _slice.ManifestStatus.Invalid, "error.default"), _obj);
var ManifestStatusIndicator = function() {
    var manifestStatus = (0, _hooks.useSelector)(_slice.getManifestStatus);
    if (manifestStatus === _slice.ManifestStatus.Unknown) {
        return null;
    }
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
        position: "absolute",
        bottom: "0",
        right: "0",
        width: "10px",
        height: "10px",
        background: MANIFEST_COLORS[manifestStatus],
        borderRadius: "5px"
    });
};
