// @metamask/snaps-simulator/dist/esm/utils/render.js
import { jsx as _jsx } from "react/jsx-runtime";
import { render as testingLibraryRender } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Root } from "../components";
import { createStore } from "../store";
/**
 * Render a component for testing. This function wraps the component in a
 * `Root` component, which provides the component with the same context that
 * it would have in the app.
 *
 * @param component - The component to render.
 * @param route - The route to render the component at. Defaults to `/`.
 * @returns The rendered component.
 */ export function render(component) {
    var route = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : "/";
    var store = createStore();
    return testingLibraryRender(component, {
        wrapper: function(param) {
            var children = param.children;
            return /*#__PURE__*/ _jsx(Root, {
                store: store,
                children: /*#__PURE__*/ _jsx(MemoryRouter, {
                    initialEntries: [
                        route
                    ],
                    initialIndex: 0,
                    children: children
                })
            });
        }
    });
}
