// @metamask/snaps-simulator/dist/cjs/features/navigation/Navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Navigation", {
    enumerable: true,
    get: function() {
        return Navigation;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _components = require("../../components");
var _hooks = require("../../hooks");
var _components1 = require("./components");
var _NavigationTag = require("./components/NavigationTag");
var _items = require("./items");
var Navigation = function() {
    var applicationState = (0, _hooks.useSelector)(function(state) {
        return state;
    });
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Container, {
        as: "nav",
        size: "fullWidth",
        display: "flex",
        flexDirection: "column",
        padding: "2",
        flex: "1",
        children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Stack, {
            as: _react.List,
            spacing: "2",
            flex: "1",
            children: [
                _items.NAVIGATION_ITEMS.map(function(param) {
                    var condition = param.condition, icon = param.icon, label = param.label, tag = param.tag, description = param.description, path = param.path;
                    if (condition && !condition(applicationState)) {
                        return null;
                    }
                    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_components1.Item, {
                        path: path,
                        tag: tag,
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                                icon: icon
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
                                children: [
                                    /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Text, {
                                        children: [
                                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                                                as: "span",
                                                fontWeight: "600",
                                                children: label
                                            }),
                                            " ",
                                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_NavigationTag.NavigationTag, {
                                                path: path,
                                                children: tag
                                            })
                                        ]
                                    }),
                                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                        fontSize: "sm",
                                        marginTop: "1",
                                        children: description
                                    })
                                ]
                            })
                        ]
                    }, path);
                }),
                /*#__PURE__*/ (0, _jsxruntime.jsxs)(_components1.Item, {
                    tag: "manifest",
                    path: "/manifest",
                    children: [
                        /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
                            position: "relative",
                            children: [
                                /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                                    icon: "manifest"
                                }),
                                /*#__PURE__*/ (0, _jsxruntime.jsx)(_components1.ManifestStatusIndicator, {})
                            ]
                        }),
                        /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
                            children: [
                                /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Text, {
                                    children: [
                                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                                            as: "span",
                                            fontWeight: "600",
                                            children: "Manifest"
                                        }),
                                        " ",
                                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_NavigationTag.NavigationTag, {
                                            path: "/manifest",
                                            children: "snap.manifest.json"
                                        })
                                    ]
                                }),
                                /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                    fontSize: "sm",
                                    marginTop: "1",
                                    children: "Validate the snap manifest"
                                })
                            ]
                        })
                    ]
                }, "manifest")
            ]
        })
    });
};
