// @metamask/snaps-simulator/dist/cjs/features/handlers/components/Handler.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Handler", {
    enumerable: true,
    get: function() {
        return Handler;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _react1 = require("react");
var _reactrouterdom = require("react-router-dom");
var _components = require("../../../components");
var _hooks = require("../../../hooks");
var _utils = require("../../../utils");
var _simulation = require("../../simulation");
var _History = require("./History");
var _PlayButton = require("./PlayButton");
var _ResetTab = require("./ResetTab");
var _ResetUserInterfaceTab = require("./ResetUserInterfaceTab");
var _Response = require("./Response");
var _UserInterface = require("./UserInterface");
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
var Handler = function() {
    var _useState = _sliced_to_array((0, _react1.useState)(0), 2), tab = _useState[0], setTab = _useState[1];
    var userInterface = (0, _hooks.useSelector)(_simulation.getUserInterface);
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Flex, {
        width: "100%",
        direction: "column",
        overflow: "hidden",
        children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
            direction: "row",
            flex: "1",
            overflow: "hidden",
            children: [
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.TestConditional, {
                    isTest: false,
                    children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Flex, {
                        direction: "column",
                        flex: "1",
                        width: "50%",
                        overflow: "hidden",
                        children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Tabs, {
                            display: "flex",
                            flexDirection: "column",
                            flex: "1",
                            overflow: "hidden",
                            isLazy: true,
                            onChange: setTab,
                            children: [
                                /*#__PURE__*/ (0, _jsxruntime.jsx)(_ResetTab.ResetTab, {}),
                                /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.TabList, {
                                    alignItems: "center",
                                    children: [
                                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Tab, {
                                            children: "Request"
                                        }),
                                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Tab, {
                                            children: "History"
                                        }),
                                        tab === 0 && /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                                            marginLeft: "auto",
                                            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_PlayButton.PlayButton, {})
                                        })
                                    ]
                                }),
                                /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.TabPanels, {
                                    display: "flex",
                                    flexDirection: "column",
                                    flex: "1",
                                    overflow: "hidden",
                                    children: [
                                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.TabPanel, {
                                            display: "flex",
                                            flexDirection: "column",
                                            flex: "1",
                                            overflowY: "auto",
                                            "data-testid": "tab-request",
                                            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Outlet, {})
                                        }),
                                        /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.TabPanel, {
                                            padding: "0",
                                            display: "flex",
                                            flexDirection: "column",
                                            flex: "1",
                                            overflowY: "auto",
                                            children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_History.History, {})
                                        })
                                    ]
                                })
                            ]
                        })
                    })
                }),
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                    display: "flex",
                    flexDirection: "column",
                    flex: "1",
                    width: "50%",
                    borderLeft: _utils.IS_TEST_BUILD ? "none" : "1px solid",
                    borderColor: "border.default",
                    children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Tabs, {
                        display: "flex",
                        flexDirection: "column",
                        flex: "1",
                        overflow: "hidden",
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_ResetTab.ResetTab, {}),
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_ResetUserInterfaceTab.ResetUserInterfaceTab, {}),
                            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.TabList, {
                                children: [
                                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Tab, {
                                        children: "Response"
                                    }),
                                    userInterface && /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Tab, {
                                        children: "UI"
                                    })
                                ]
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.TabPanels, {
                                display: "flex",
                                flexDirection: "column",
                                flex: "1",
                                overflow: "hidden",
                                children: [
                                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.TabPanel, {
                                        display: "flex",
                                        flexDirection: "column",
                                        flex: "1",
                                        padding: "0",
                                        "data-testid": "tab-response",
                                        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_Response.Response, {})
                                    }),
                                    userInterface && /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.TabPanel, {
                                        overflowY: "auto",
                                        "data-testid": "tab-ui",
                                        children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_UserInterface.UserInterface, {})
                                    })
                                ]
                            })
                        ]
                    })
                })
            ]
        })
    });
};
