// @metamask/snaps-simulator/dist/cjs/features/handlers/components/HistoryItem.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "HistoryItem", {
    enumerable: true,
    get: function() {
        return HistoryItem;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _datefns = require("date-fns");
var _components = require("../../../components");
var _hooks = require("../../../hooks");
var HistoryItem = function(param) {
    var item = param.item;
    var _item_request_request, _item_request, _item_request1;
    var dispatch = (0, _hooks.useDispatch)();
    var tab = (0, _react.useTabsContext)();
    var handler = (0, _hooks.useHandler)();
    var handleClick = function() {
        dispatch({
            type: "".concat(handler, "/setRequestFromHistory"),
            payload: item.request
        });
        tab.setSelectedIndex(0);
    };
    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.ListItem, {
        onClick: handleClick,
        children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Popover, {
            trigger: "hover",
            openDelay: 10,
            children: [
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.PopoverTrigger, {
                    children: /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
                        margin: "2",
                        padding: "2",
                        borderRadius: "md",
                        justifyContent: "space-between",
                        alignItems: "center",
                        _hover: {
                            cursor: "pointer",
                            background: "background.hover"
                        },
                        gap: "2",
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Box, {
                                overflow: "hidden",
                                children: [
                                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                        fontSize: "sm",
                                        fontWeight: "600",
                                        overflow: "hidden",
                                        textOverflow: "ellipsis",
                                        whiteSpace: "nowrap",
                                        children: (_item_request = item.request) === null || _item_request === void 0 ? void 0 : (_item_request_request = _item_request.request) === null || _item_request_request === void 0 ? void 0 : _item_request_request.method
                                    }),
                                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                                        color: "text.muted",
                                        fontSize: "sm",
                                        children: (0, _datefns.formatDistance)(item.date, new Date(), {
                                            addSuffix: true
                                        })
                                    })
                                ]
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Box, {
                                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Icon, {
                                    icon: "arrowTopRight",
                                    width: "10px"
                                })
                            })
                        ]
                    })
                }),
                /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.PopoverContent, {
                    padding: "4",
                    fontFamily: "code",
                    fontSize: "xs",
                    overflow: "hidden",
                    textOverflow: "ellipsis",
                    whiteSpace: "nowrap",
                    children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Text, {
                        display: "block",
                        as: "pre",
                        overflow: "hidden",
                        textOverflow: "ellipsis",
                        whiteSpace: "pre",
                        children: JSON.stringify((_item_request1 = item.request) === null || _item_request1 === void 0 ? void 0 : _item_request1.request, null, 2)
                    })
                })
            ]
        })
    });
};
