// @metamask/snaps-simulator/dist/esm/features/simulation/sagas.js
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_without_holes(arr) {
    if (Array.isArray(arr)) return _array_like_to_array(arr);
}
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _iterable_to_array(iter) {
    if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);
}
function _non_iterable_spread() {
    throw new TypeError("Invalid attempt to spread non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function _to_consumable_array(arr) {
    return _array_without_holes(arr) || _iterable_to_array(arr) || _unsupported_iterable_to_array(arr) || _non_iterable_spread();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
import { ControllerMessenger } from "@metamask/base-controller";
import { encrypt, decrypt } from "@metamask/browser-passworder";
import { createFetchMiddleware } from "@metamask/eth-json-rpc-middleware";
import { mnemonicPhraseToBytes } from "@metamask/key-tree";
import { PermissionController, SubjectMetadataController, SubjectType } from "@metamask/permission-controller";
import { serializeError } from "@metamask/rpc-errors";
import { caveatSpecifications as snapsCaveatsSpecifications } from "@metamask/rpc-methods";
import { IframeExecutionService, setupMultiplex, endowmentCaveatSpecifications as snapsEndowmentCaveatSpecifications, processSnapPermissions, buildSnapEndowmentSpecifications, buildSnapRestrictedMethodSpecifications } from "@metamask/snaps-controllers";
import packageJson from "@metamask/snaps-execution-environments/package.json";
import { logError } from "@metamask/snaps-utils";
import { getSafeJson } from "@metamask/utils";
import { JsonRpcEngine } from "json-rpc-engine";
import { createEngineStream } from "json-rpc-middleware-stream";
import pump from "pump";
import { all, call, put, select, takeLatest } from "redux-saga/effects";
import { runSaga } from "../../store/middleware";
import { getSnapId, getSrp, setSnapId } from "../configuration";
import { addError } from "../console";
import { ManifestStatus, setValid } from "../manifest";
import { JSON_RPC_ENDPOINT } from "./constants";
import { getSnapState, showDialog, showInAppNotification, showNativeNotification, updateSnapState } from "./hooks";
import { createMiscMethodMiddleware } from "./middleware";
import { getExecutionService, setExecutionService, setSourceCode, sendRequest, SnapStatus, setStatus, getPermissionController, setManifest, setPermissionController, setSubjectMetadataController, getSubjectMetadataController } from "./slice";
import { ExcludedSnapEndowments, getEndowments, unrestrictedMethods } from "./snap-permissions";
var DEFAULT_ENVIRONMENT_URL = "https://execution.metamask.io/".concat(packageJson.version, "/index.html");
/**
 * The initialization saga is run on when the snap ID is changed and initializes the snaps execution environment.
 * This saga also creates the JSON-RPC engine and middlewares used to process RPC requests from the executing snap.
 *
 * @param action - The action itself.
 * @param action.payload - The payload of the action, in this case the snap ID.
 * @yields Puts the execution environment after creation.
 */ export function initSaga(param) {
    var payload, controllerMessenger, srp, permissionSpecifications, subjectMetadataController, permissionController, engine, searchParams, _searchParams_get, environmentUrl, executionService;
    return _ts_generator(this, function(_state) {
        switch(_state.label){
            case 0:
                payload = param.payload;
                controllerMessenger = new ControllerMessenger();
                return [
                    4,
                    select(getSrp)
                ];
            case 1:
                srp = _state.sent();
                permissionSpecifications = _object_spread({}, buildSnapEndowmentSpecifications(Object.keys(ExcludedSnapEndowments)), buildSnapRestrictedMethodSpecifications([], {
                    // TODO: Add all the hooks required
                    encrypt: encrypt,
                    decrypt: decrypt,
                    getUnlockPromise: /*#__PURE__*/ _async_to_generator(function() {
                        return _ts_generator(this, function(_state) {
                            return [
                                2,
                                Promise.resolve(true)
                            ];
                        });
                    }),
                    getMnemonic: /*#__PURE__*/ _async_to_generator(function() {
                        return _ts_generator(this, function(_state) {
                            return [
                                2,
                                mnemonicPhraseToBytes(srp)
                            ];
                        });
                    }),
                    showDialog: /*#__PURE__*/ _async_to_generator(function() {
                        var _len, args, _key;
                        var _arguments = arguments;
                        return _ts_generator(this, function(_state) {
                            switch(_state.label){
                                case 0:
                                    for(_len = _arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
                                        args[_key] = _arguments[_key];
                                    }
                                    return [
                                        4,
                                        runSaga.apply(void 0, [
                                            showDialog
                                        ].concat(_to_consumable_array(args))).toPromise()
                                    ];
                                case 1:
                                    return [
                                        2,
                                        _state.sent()
                                    ];
                            }
                        });
                    }),
                    showNativeNotification: /*#__PURE__*/ _async_to_generator(function() {
                        var _len, args, _key;
                        var _arguments = arguments;
                        return _ts_generator(this, function(_state) {
                            switch(_state.label){
                                case 0:
                                    for(_len = _arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
                                        args[_key] = _arguments[_key];
                                    }
                                    return [
                                        4,
                                        runSaga.apply(void 0, [
                                            showNativeNotification
                                        ].concat(_to_consumable_array(args))).toPromise()
                                    ];
                                case 1:
                                    return [
                                        2,
                                        _state.sent()
                                    ];
                            }
                        });
                    }),
                    showInAppNotification: /*#__PURE__*/ _async_to_generator(function() {
                        var _len, args, _key;
                        var _arguments = arguments;
                        return _ts_generator(this, function(_state) {
                            switch(_state.label){
                                case 0:
                                    for(_len = _arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
                                        args[_key] = _arguments[_key];
                                    }
                                    return [
                                        4,
                                        runSaga.apply(void 0, [
                                            showInAppNotification
                                        ].concat(_to_consumable_array(args))).toPromise()
                                    ];
                                case 1:
                                    return [
                                        2,
                                        _state.sent()
                                    ];
                            }
                        });
                    }),
                    getSnapState: /*#__PURE__*/ _async_to_generator(function() {
                        var _len, args, _key;
                        var _arguments = arguments;
                        return _ts_generator(this, function(_state) {
                            switch(_state.label){
                                case 0:
                                    for(_len = _arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
                                        args[_key] = _arguments[_key];
                                    }
                                    return [
                                        4,
                                        runSaga.apply(void 0, [
                                            getSnapState
                                        ].concat(_to_consumable_array(args))).toPromise()
                                    ];
                                case 1:
                                    return [
                                        2,
                                        _state.sent()
                                    ];
                            }
                        });
                    }),
                    updateSnapState: /*#__PURE__*/ _async_to_generator(function() {
                        var _len, args, _key;
                        var _arguments = arguments;
                        return _ts_generator(this, function(_state) {
                            switch(_state.label){
                                case 0:
                                    for(_len = _arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
                                        args[_key] = _arguments[_key];
                                    }
                                    return [
                                        4,
                                        runSaga.apply(void 0, [
                                            updateSnapState
                                        ].concat(_to_consumable_array(args))).toPromise()
                                    ];
                                case 1:
                                    return [
                                        2,
                                        _state.sent()
                                    ];
                            }
                        });
                    }),
                    clearSnapState: /*#__PURE__*/ _async_to_generator(function() {
                        var _len, args, _key;
                        var _arguments = arguments;
                        return _ts_generator(this, function(_state) {
                            switch(_state.label){
                                case 0:
                                    for(_len = _arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++){
                                        args[_key] = _arguments[_key];
                                    }
                                    return [
                                        4,
                                        runSaga(updateSnapState, args[0], null).toPromise()
                                    ];
                                case 1:
                                    return [
                                        2,
                                        _state.sent()
                                    ];
                            }
                        });
                    })
                }));
                subjectMetadataController = new SubjectMetadataController({
                    messenger: controllerMessenger.getRestricted({
                        name: "SubjectMetadataController"
                    }),
                    subjectCacheLimit: 100
                });
                permissionController = new PermissionController({
                    messenger: controllerMessenger.getRestricted({
                        name: "PermissionController",
                        allowedActions: [
                            "ApprovalController:addRequest",
                            "ApprovalController:hasRequest",
                            "ApprovalController:acceptRequest",
                            "ApprovalController:rejectRequest",
                            "SnapController:getPermitted",
                            "SnapController:install",
                            "SubjectMetadataController:getSubjectMetadata"
                        ]
                    }),
                    caveatSpecifications: _object_spread({}, snapsCaveatsSpecifications, snapsEndowmentCaveatSpecifications),
                    permissionSpecifications: permissionSpecifications,
                    unrestrictedMethods: unrestrictedMethods
                });
                engine = new JsonRpcEngine();
                engine.push(createMiscMethodMiddleware());
                engine.push(permissionController.createPermissionMiddleware({
                    origin: payload
                }));
                engine.push(createFetchMiddleware({
                    btoa: globalThis.btoa,
                    fetch: globalThis.fetch,
                    rpcUrl: JSON_RPC_ENDPOINT
                }));
                searchParams = new URLSearchParams(window.location.search);
                environmentUrl = (_searchParams_get = searchParams.get("environment")) !== null && _searchParams_get !== void 0 ? _searchParams_get : DEFAULT_ENVIRONMENT_URL;
                executionService = new IframeExecutionService({
                    iframeUrl: new URL(environmentUrl),
                    messenger: controllerMessenger.getRestricted({
                        name: "ExecutionService"
                    }),
                    setupSnapProvider: function(_snapId, rpcStream) {
                        var mux = setupMultiplex(rpcStream, "snapStream");
                        var stream = mux.createStream("metamask-provider");
                        var providerStream = createEngineStream({
                            engine: engine
                        });
                        pump(stream, providerStream, stream);
                    }
                });
                return [
                    4,
                    put(setExecutionService(executionService))
                ];
            case 2:
                _state.sent();
                return [
                    4,
                    put(setPermissionController(permissionController))
                ];
            case 3:
                _state.sent();
                return [
                    4,
                    put(setSubjectMetadataController(subjectMetadataController))
                ];
            case 4:
                _state.sent();
                return [
                    2
                ];
        }
    });
}
/**
 * The reboot saga, which should run when the setSourceCode action is emitted.
 * This saga will terminate existing snaps and reboot the snap with the latest source code.
 *
 * @param action - The action itself.
 * @param action.payload - The payload of the action, in this case the source code.
 * @yields Select for selecting the execution service and call to call the execution service.
 */ export function rebootSaga(param) {
    var payload, snapId, executionService, permissionController, endowments, error;
    return _ts_generator(this, function(_state) {
        switch(_state.label){
            case 0:
                payload = param.payload;
                return [
                    4,
                    select(getSnapId)
                ];
            case 1:
                snapId = _state.sent();
                return [
                    4,
                    select(getExecutionService)
                ];
            case 2:
                executionService = _state.sent();
                return [
                    4,
                    select(getPermissionController)
                ];
            case 3:
                permissionController = _state.sent();
                return [
                    4,
                    call(getEndowments, permissionController, snapId)
                ];
            case 4:
                endowments = _state.sent();
                _state.label = 5;
            case 5:
                _state.trys.push([
                    5,
                    9,
                    ,
                    12
                ]);
                return [
                    4,
                    call([
                        executionService,
                        "terminateAllSnaps"
                    ])
                ];
            case 6:
                _state.sent();
                return [
                    4,
                    call([
                        executionService,
                        "executeSnap"
                    ], {
                        snapId: snapId,
                        sourceCode: payload.toString("utf8"),
                        endowments: endowments
                    })
                ];
            case 7:
                _state.sent();
                return [
                    4,
                    put(setStatus(SnapStatus.Ok))
                ];
            case 8:
                _state.sent();
                return [
                    3,
                    12
                ];
            case 9:
                error = _state.sent();
                logError(error);
                return [
                    4,
                    put(addError(error))
                ];
            case 10:
                _state.sent();
                return [
                    4,
                    put(setStatus(SnapStatus.Error))
                ];
            case 11:
                _state.sent();
                return [
                    3,
                    12
                ];
            case 12:
                return [
                    2
                ];
        }
    });
}
/**
 * The request saga, which should run when the sendRequest action is emitted.
 * This saga will send an RPC request to the snap and capture the response.
 *
 * @param action - The action itself.
 * @param action.payload - The payload of the action, in this case the RPC request.
 * @yields Select for selecting the execution service, call to call the execution service and put for storing the response.
 */ export function requestSaga(param) {
    var payload, snapId, executionService, result, error;
    return _ts_generator(this, function(_state) {
        switch(_state.label){
            case 0:
                payload = param.payload;
                return [
                    4,
                    put({
                        type: "".concat(payload.handler, "/setRequest"),
                        payload: payload
                    })
                ];
            case 1:
                _state.sent();
                return [
                    4,
                    select(getSnapId)
                ];
            case 2:
                snapId = _state.sent();
                return [
                    4,
                    select(getExecutionService)
                ];
            case 3:
                executionService = _state.sent();
                _state.label = 4;
            case 4:
                _state.trys.push([
                    4,
                    7,
                    ,
                    9
                ]);
                return [
                    4,
                    call([
                        executionService,
                        "handleRpcRequest"
                    ], snapId, payload)
                ];
            case 5:
                result = _state.sent();
                return [
                    4,
                    put({
                        type: "".concat(payload.handler, "/setResponse"),
                        payload: {
                            result: result
                        }
                    })
                ];
            case 6:
                _state.sent();
                return [
                    3,
                    9
                ];
            case 7:
                error = _state.sent();
                return [
                    4,
                    put({
                        type: "".concat(payload.handler, "/setResponse"),
                        payload: {
                            error: serializeError(error)
                        }
                    })
                ];
            case 8:
                _state.sent();
                return [
                    3,
                    9
                ];
            case 9:
                return [
                    2
                ];
        }
    });
}
/**
 * The permissions saga, which should run when the setManifest action is emitted.
 * This saga will automatically grant the active snap all permissions defined in the snap manifest.
 *
 * @param action - The action itself.
 * @param action.payload - The payload of the action, in this case a snap manifest.
 * @yields Selects the permission controller
 */ export function permissionsSaga(param) {
    var payload, snapId, subjectMetadataController, permissionController, approvedPermissions, error;
    return _ts_generator(this, function(_state) {
        switch(_state.label){
            case 0:
                payload = param.payload;
                _state.label = 1;
            case 1:
                _state.trys.push([
                    1,
                    7,
                    ,
                    11
                ]);
                return [
                    4,
                    select(getSnapId)
                ];
            case 2:
                snapId = _state.sent();
                return [
                    4,
                    select(getSubjectMetadataController)
                ];
            case 3:
                subjectMetadataController = _state.sent();
                return [
                    4,
                    call([
                        subjectMetadataController,
                        "addSubjectMetadata"
                    ], {
                        origin: snapId,
                        subjectType: SubjectType.Snap
                    })
                ];
            case 4:
                _state.sent();
                return [
                    4,
                    select(getPermissionController)
                ];
            case 5:
                permissionController = _state.sent();
                approvedPermissions = processSnapPermissions(getSafeJson(payload.result.initialPermissions));
                // Grant all permissions
                return [
                    4,
                    call([
                        permissionController,
                        "grantPermissions"
                    ], {
                        approvedPermissions: approvedPermissions,
                        subject: {
                            origin: snapId
                        },
                        preserveExistingPermissions: false
                    })
                ];
            case 6:
                _state.sent();
                return [
                    3,
                    11
                ];
            case 7:
                error = _state.sent();
                logError(error);
                return [
                    4,
                    put(addError(error))
                ];
            case 8:
                _state.sent();
                return [
                    4,
                    put(setStatus(SnapStatus.Error))
                ];
            case 9:
                _state.sent();
                return [
                    4,
                    put(setValid(ManifestStatus.Unknown))
                ];
            case 10:
                _state.sent();
                return [
                    3,
                    11
                ];
            case 11:
                return [
                    2
                ];
        }
    });
}
/**
 * The root simulation saga which runs all sagas in this file.
 *
 * @yields All sagas for the simulation feature.
 */ export function simulationSaga() {
    return _ts_generator(this, function(_state) {
        switch(_state.label){
            case 0:
                return [
                    4,
                    all([
                        takeLatest(setSnapId.type, initSaga),
                        takeLatest(setSourceCode.type, rebootSaga),
                        takeLatest(sendRequest.type, requestSaga),
                        takeLatest(setManifest, permissionsSaga)
                    ])
                ];
            case 1:
                _state.sent();
                return [
                    2
                ];
        }
    });
}
