// @metamask/snaps-simulator/dist/esm/features/console/Console.js
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex, Tab, TabList, TabPanel, TabPanels, Tabs } from "@chakra-ui/react";
import { useEffect, useRef, useState } from "react";
import { Icon } from "../../components";
import { useSelector } from "../../hooks";
import { ConsoleContent } from "./ConsoleContent";
import { getConsoleEntries } from "./slice";
/**
 * Console component.
 *
 * @returns The console component.
 */ export var Console = function() {
    var ref = useRef(null);
    var _useState = _sliced_to_array(useState(false), 2), collapsed = _useState[0], setCollapsed = _useState[1];
    var entries = useSelector(getConsoleEntries);
    useEffect(function() {
        if (ref.current) {
            // TODO: Maybe not scroll if the user is scrolled up?
            ref.current.scrollTop = ref.current.scrollHeight;
        }
    }, [
        entries
    ]);
    var handleToggleConsole = function() {
        setCollapsed(function(state) {
            return !state;
        });
    };
    return /*#__PURE__*/ _jsx(Flex, {
        borderTop: "1px solid",
        borderColor: "border.default",
        height: collapsed ? "47px" : "266px",
        transition: "height 0.5s",
        flexDirection: "column",
        overflow: "hidden",
        children: /*#__PURE__*/ _jsxs(Tabs, {
            display: "flex",
            flexDirection: "column",
            flex: "1",
            overflow: "hidden",
            children: [
                /*#__PURE__*/ _jsxs(TabList, {
                    pr: "2",
                    children: [
                        /*#__PURE__*/ _jsx(Tab, {
                            children: "Console"
                        }),
                        /*#__PURE__*/ _jsx(Flex, {
                            alignItems: "center",
                            ml: "auto",
                            onClick: handleToggleConsole,
                            _hover: {
                                cursor: "pointer"
                            },
                            children: /*#__PURE__*/ _jsx(Icon, {
                                icon: "arrowDown",
                                height: "14px",
                                transform: collapsed ? "rotate(180deg)" : "rotate(0deg)"
                            })
                        })
                    ]
                }),
                /*#__PURE__*/ _jsx(TabPanels, {
                    display: "flex",
                    flexDirection: "column",
                    flex: "1",
                    overflow: "hidden",
                    children: /*#__PURE__*/ _jsx(TabPanel, {
                        ref: ref,
                        display: "flex",
                        flexDirection: "column",
                        flex: "1",
                        overflowY: "auto",
                        children: /*#__PURE__*/ _jsx(ConsoleContent, {})
                    })
                })
            ]
        })
    });
};
