// @metamask/snaps-simulator/dist/esm/features/console/slice.js
import { createSelector, createSlice } from "@reduxjs/toolkit";
export var ConsoleEntryType;
(function(ConsoleEntryType) {
    ConsoleEntryType["Log"] = "log";
    ConsoleEntryType["Error"] = "error";
})(ConsoleEntryType || (ConsoleEntryType = {}));
var INITIAL_STATE = {
    entries: []
};
var slice = createSlice({
    name: "console",
    initialState: INITIAL_STATE,
    reducers: {
        addDefault: function addDefault(state, action) {
            state.entries.push({
                date: Date.now(),
                type: ConsoleEntryType.Log,
                message: action.payload
            });
        },
        addError: function addError(state, action) {
            var _action_payload_stack;
            state.entries.push({
                date: Date.now(),
                type: ConsoleEntryType.Error,
                message: (_action_payload_stack = action.payload.stack) !== null && _action_payload_stack !== void 0 ? _action_payload_stack : action.payload.message
            });
        }
    }
});
var _slice_actions = slice.actions;
export var addDefault = _slice_actions.addDefault, addError = _slice_actions.addError;
export var console = slice.reducer;
export var getConsoleEntries = createSelector(function(state) {
    return state.console;
}, function(state) {
    return state.entries;
});
