// @metamask/snaps-simulator/dist/cjs/features/layout/Layout.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Layout", {
    enumerable: true,
    get: function() {
        return Layout;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _reactrouterdom = require("react-router-dom");
var _console = require("../console");
var _components = require("./components");
var Layout = function() {
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
        direction: "column",
        height: "100vh",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Header, {}),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
                direction: "row",
                flex: "1",
                overflow: "hidden",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Sidebar, {}),
                    /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Flex, {
                        direction: "column",
                        width: "full",
                        children: [
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Flex, {
                                flex: "1",
                                overflow: "hidden",
                                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_reactrouterdom.Outlet, {})
                            }),
                            /*#__PURE__*/ (0, _jsxruntime.jsx)(_console.Console, {})
                        ]
                    })
                ]
            })
        ]
    });
};
