// @metamask/snaps-simulator/dist/cjs/features/polling/sagas.test.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _jestfetchmock = /*#__PURE__*/ _interop_require_default(require("jest-fetch-mock"));
var _reduxsagatestplan = require("redux-saga-test-plan");
var _simulation = require("../simulation");
var _mockManifest = require("../simulation/test/mockManifest");
var _mockSnap = require("../simulation/test/mockSnap");
var _sagas = require("./sagas");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _ts_generator(thisArg, body) {
    var f, y, t, g, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    };
    return g = {
        next: verb(0),
        "throw": verb(1),
        "return": verb(2)
    }, typeof Symbol === "function" && (g[Symbol.iterator] = function() {
        return this;
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(_)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
_jestfetchmock.default.enableMocks();
describe("pollingSaga", function() {
    it("calls the fetching saga and delay for local snaps", /*#__PURE__*/ _async_to_generator(function() {
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    return [
                        4,
                        (0, _reduxsagatestplan.expectSaga)(_sagas.pollingSaga).withState({
                            configuration: {
                                snapId: "local:http://localhost:8080"
                            }
                        }).call(_sagas.fetchingSaga).delay(500).silentRun()
                    ];
                case 1:
                    _state.sent();
                    return [
                        2
                    ];
            }
        });
    }));
    it("calls the fetching saga and not delay for npm snaps", /*#__PURE__*/ _async_to_generator(function() {
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    return [
                        4,
                        (0, _reduxsagatestplan.expectSaga)(_sagas.pollingSaga).withState({
                            configuration: {
                                snapId: "npm:@metamask/test-snap-bip32"
                            }
                        }).call(_sagas.fetchingSaga).not.delay(500).silentRun()
                    ];
                case 1:
                    _state.sent();
                    return [
                        2
                    ];
            }
        });
    }));
});
describe("fetchingSaga", function() {
    beforeEach(function() {
        _jestfetchmock.default.resetMocks();
    });
    it("fetches the snap and updates manifest and source code", /*#__PURE__*/ _async_to_generator(function() {
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    _jestfetchmock.default.mockResponses(JSON.stringify(_mockManifest.MOCK_MANIFEST), _mockSnap.MOCK_SNAP_SOURCE, _mockSnap.MOCK_SNAP_ICON);
                    return [
                        4,
                        (0, _reduxsagatestplan.expectSaga)(_sagas.fetchingSaga).withState({
                            configuration: {
                                snapId: "local:http://localhost:8080"
                            },
                            simulation: {
                                manifest: null
                            }
                        }).put((0, _simulation.setManifest)(_mockManifest.MOCK_MANIFEST_FILE)).put((0, _simulation.setSourceCode)(_mockSnap.MOCK_SNAP_SOURCE_FILE)).put((0, _simulation.setIcon)(_mockSnap.MOCK_SNAP_ICON_FILE)).silentRun()
                    ];
                case 1:
                    _state.sent();
                    return [
                        2
                    ];
            }
        });
    }));
    it("doesnt update the source code if checksum matches cached checksum", /*#__PURE__*/ _async_to_generator(function() {
        return _ts_generator(this, function(_state) {
            switch(_state.label){
                case 0:
                    _jestfetchmock.default.mockResponses(JSON.stringify(_mockManifest.MOCK_MANIFEST));
                    return [
                        4,
                        (0, _reduxsagatestplan.expectSaga)(_sagas.fetchingSaga).withState({
                            configuration: {
                                snapId: "local:http://localhost:8080"
                            },
                            simulation: {
                                manifest: _mockManifest.MOCK_MANIFEST_FILE
                            }
                        }).not.put.actionType(_simulation.setManifest.type).not.put.actionType(_simulation.setSourceCode.type).not.put.actionType(_simulation.setIcon.type).silentRun()
                    ];
                case 1:
                    _state.sent();
                    return [
                        2
                    ];
            }
        });
    }));
});
