// @metamask/snaps-simulator/dist/esm/components/dialogs/AlertDialog.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Button, Flex } from "@chakra-ui/react";
import { Renderer } from "../../features/renderer";
import { Delineator } from "../Delineator";
import { Window } from "../Window";
/**
 * Snap alert dialog.
 *
 * @param props - The component props.
 * @param props.snapName - The snap name.
 * @param props.snapId - The snap ID.
 * @param props.node - The component to render.
 * @param props.onClose - The close callback.
 * @returns The component.
 */ export var AlertDialog = function(param) {
    var snapName = param.snapName, snapId = param.snapId, node = param.node, onClose = param.onClose;
    return /*#__PURE__*/ _jsxs(Window, {
        snapName: snapName,
        snapId: snapId,
        children: [
            /*#__PURE__*/ _jsx(Box, {
                margin: "4",
                marginTop: "0",
                flex: "1",
                children: /*#__PURE__*/ _jsx(Delineator, {
                    snapName: snapName,
                    children: /*#__PURE__*/ _jsx(Renderer, {
                        node: node
                    })
                })
            }),
            /*#__PURE__*/ _jsx(Flex, {
                borderTop: "1px solid",
                borderTopColor: "border.default",
                paddingTop: "4",
                paddingX: "4",
                gap: "4",
                children: /*#__PURE__*/ _jsx(Button, {
                    variant: "primary",
                    flex: "1",
                    onClick: onClose,
                    children: "OK"
                })
            })
        ]
    });
};
