// @airstack/node/dist/cjs/utils/query/createVariable.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
exports.createVariable = void 0;
var graphql_1 = require("graphql");
function getType(kind, value, required) {
    if (required === void 0) { required = false; }
    var type = {
        kind: "NamedType",
        name: {
            kind: "Name",
            value: value
        }
    };
    if (required) {
        type = {
            kind: "NonNullType",
            type: __assign({}, type)
        };
    }
    if (kind === graphql_1.Kind.LIST) {
        type = {
            kind: "ListType",
            type: __assign({}, type)
        };
    }
    return type;
}
function createVariable(_a) {
    var type = _a.type, isRequired = _a.isRequired, name = _a.name, uniqueName = _a.uniqueName, valueKind = _a.valueKind;
    return {
        kind: "VariableDefinition",
        variable: {
            kind: "Variable",
            name: {
                kind: "Name",
                value: uniqueName || name
            }
        },
        type: getType(valueKind, type, isRequired),
        directives: []
    };
}
exports.createVariable = createVariable;
