// @metamask/snaps-simulator/dist/cjs/features/layout/components/Header.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Header", {
    enumerable: true,
    get: function() {
        return Header;
    }
});
var _jsxruntime = require("react/jsx-runtime");
var _react = require("@chakra-ui/react");
var _components = require("../../../components");
var _configuration = require("../../configuration");
var _status = require("../../status");
var Header = function() {
    return /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Container, {
        as: "header",
        size: "fullWidth",
        borderBottom: "muted",
        display: "flex",
        flexDirection: "row",
        justifyContent: "space-between",
        children: [
            /*#__PURE__*/ (0, _jsxruntime.jsx)(_react.Stack, {
                direction: "row",
                height: "7",
                align: "center",
                children: /*#__PURE__*/ (0, _jsxruntime.jsx)(_components.Logo, {})
            }),
            /*#__PURE__*/ (0, _jsxruntime.jsxs)(_react.Stack, {
                direction: "row",
                height: "7",
                align: "center",
                children: [
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_status.StatusIndicator, {}),
                    /*#__PURE__*/ (0, _jsxruntime.jsx)(_configuration.Configuration, {})
                ]
            })
        ]
    });
};
