// @metamask/snaps-simulator/dist/esm/features/status/StatusIndicator.js
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Flex, Spinner, Text } from "@chakra-ui/react";
import { SnapIdPrefixes, getSnapPrefix, stripSnapPrefix } from "@metamask/snaps-utils";
import { useEffect, useState } from "react";
import { Icon } from "../../components";
import { useDispatch, useSelector } from "../../hooks";
import { getSnapId, openConfigurationModal } from "../configuration";
import { SnapStatus, getStatus } from "../simulation";
export var StatusIndicator = function() {
    var snapId = useSelector(getSnapId);
    var status = useSelector(getStatus);
    var dispatch = useDispatch();
    var _useState = _sliced_to_array(useState(snapId), 2), prettyUrl = _useState[0], setPrettyUrl = _useState[1];
    useEffect(function() {
        if (!snapId) {
            return;
        }
        try {
            var stripped = stripSnapPrefix(snapId);
            if (getSnapPrefix(snapId) === SnapIdPrefixes.npm) {
                setPrettyUrl(stripped);
                return;
            }
            var url = new URL(stripped);
            setPrettyUrl(url.host);
        } catch (e) {
        // Ignore.
        }
    }, [
        snapId
    ]);
    var color = // eslint-disable-next-line no-nested-ternary
    status === SnapStatus.Ok ? "text.success" : status === SnapStatus.Error ? "text.error" : "info.default";
    var handleClick = function() {
        dispatch(openConfigurationModal());
    };
    return /*#__PURE__*/ _jsxs(Flex, {
        alignItems: "center",
        gap: "2",
        onClick: handleClick,
        cursor: "pointer",
        _hover: {
            p: {
                color: color
            }
        },
        children: [
            status === SnapStatus.Ok && /*#__PURE__*/ _jsx(Icon, {
                icon: "dot",
                width: "8px",
                "data-testid": "status-ok"
            }),
            status === SnapStatus.Loading && /*#__PURE__*/ _jsx(Spinner, {
                color: color,
                size: "xs"
            }),
            status === SnapStatus.Error && /*#__PURE__*/ _jsx(Icon, {
                icon: "errorTriangle",
                width: "16px",
                "data-testid": "status-error"
            }),
            /*#__PURE__*/ _jsx(Text, {
                fontWeight: "500",
                fontSize: "sm",
                color: "text.muted",
                transition: "color ease 0.2s",
                children: prettyUrl
            })
        ]
    });
};
